// nav-frontend-spinner/src/spinner.tsx
import * as PT from 'prop-types';
import * as React from 'react';
import * as classNames from 'classnames';
import 'nav-frontend-spinner-style';

const cls = (className, storrelse) => classNames('spinner', className, {
    [`spinner--${storrelse.toLowerCase()}`]: !!storrelse
});

export interface NavFrontendSpinnerProps {
    className?: string;
    'aria-label'?: string;
}

export interface NavFrontendSpinnerBaseProps extends NavFrontendSpinnerProps {
    type?: 'L' | 'XL' | 'XXL' | 'XXXL';
}

class NavFrontendSpinner extends React.Component<NavFrontendSpinnerBaseProps> {
    render() {
        const {type, className, ...props} = this.props;

        return (
            <div className={cls(className, type)} {...props}>
                <div/>
                <div/>
                <div/>
                <div/>
            </div>
        );
    }
}

(NavFrontendSpinner as React.ComponentClass).propTypes = {
    type: PT.oneOf(['L', 'XL', 'XXL', 'XXXL']),
    className: PT.string,
    'aria-label': PT.string
};

(NavFrontendSpinner as React.ComponentClass).defaultProps = {
    'aria-label': 'Laster innhold',
    type: 'L',
    className: undefined
};

export default NavFrontendSpinner;
