// nav-frontend-js-utils/src/eventThrottler.js
import PT from 'prop-types';
import { Component } from 'react';
import _throttle from 'lodash.throttle';

export class EventThrottler extends Component { // eslint-disable-line import/prefer-default-export
    componentDidMount() {
        window.addEventListener(this.props.event, this.throttleFn()());
    }

    componentWillUnmount() {
        if (this.actualThrottleFn) {
            window.removeEventListener(this.props.event, this.actualThrottleFn);
        }
    }

    throttleFn() {
        let cb = this.props.callback;
        cb = cb.bind(this.props.context); // eslint-disable-line react/prop-types
        this.actualThrottleFn = () => _throttle(cb, this.props.delay, { leading: false });
        return this.actualThrottleFn;
    }

    render() {
        return this.props.component;
    }
}

EventThrottler.propTypes = {
    component: PT.element.isRequired,
    callback: PT.func.isRequired,
    delay: PT.number,
    event: PT.oneOf(['resize'])
};

EventThrottler.defaultProps = {
    callback: () => {},
    context: {},
    delay: 0,
    event: 'resize'
};
