// nav-frontend-js-utils/src/index.ts
export function autobind(ctx): void {
    Object.getOwnPropertyNames(ctx.constructor.prototype)
        .filter((prop) => typeof ctx[prop] === 'function')
        .forEach((method) => {
            // eslint-disable-next-line
            ctx[method] = ctx[method].bind(ctx);
        });
}

function s4(): string {
    return Math.floor((1 + Math.random()) * 0x10000)
        .toString()
        .substring(1);
}

export function guid(): string {
    return `${s4()}${s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`;
}

export function fn(value) {
    return typeof value === 'function' ? value : () => value;
}

export function addPropToObject(obj: object, [key, value]: [string|number, any]): object {
    obj[key] = value; // eslint-disable-line no-param-reassign
    return obj;
}

export function omit(obj: object, ...props) {
    return Object.entries(obj)
        .filter(([key]) => !props.includes(key))
        .reduce(addPropToObject, {});
}

function nativeRaf() {
    if (window) {
        return window.requestAnimationFrame || window.webkitRequestAnimationFrame;
    }
    return undefined;
}

function polyfillRaf() {
    return (callback, element, delay) => {
        setTimeout(callback, delay || (1000 / 60), new Date().getTime());
    };
}

export const requestAnimationFrame = nativeRaf() || polyfillRaf();

export function clamp(lower: number, upper: number, numberish) {
    const number = parseFloat(numberish);
    if (number < lower) return lower;
    if (number > upper) return upper;
    return number;
}

export { EventThrottler } from './eventThrottler';
