// nav-frontend-popover/src/popover-base.tsx
import * as React from 'react';
import * as PT from 'prop-types';
import * as classnames from 'classnames';

import 'nav-frontend-popover-style';

const cls = (props) => classnames('popover', props.className, {
    'popover--apen': props.apen,
    'popover--over': props.orientering === 'over',
    'popover--under': props.orientering === 'under',
    'popover--venstre': props.orientering === 'venstre',
    'popover--høyre': props.orientering === 'høyre',
    'popover--under-høyre': props.orientering === 'under-høyre',
    'popover--under-venstre': props.orientering === 'under-venstre',
    'popover--uten-pil': props.utenPil
});

export interface PopoverBaseProps {
    /**
     * Egendefinert klassenavn.
     */
    className?: string;
    /**
     * Bestemmer om popover skal vises eller skjules (default)
     */
    apen: boolean;
    /**
     * Bredde på popover-vindu. Valgfri, men anbefalt å angi da posisjonering av
     * popover vindu kan bli bittelitt unøyaktig uten.
     */
    width?: number;
    /**
     * Valgfri fast høyde på popover-vindu. Hvis denne ikke angis så vil høyden på vinduet skalere
     * fritt avhengig av innholdet.
     */
    height?: number;
    /**
     * Bestemmer hvordan popover-vinduet skal posisjoneres i forhold til anker-elementet.
     */
    orientering?: 'over' | 'under' | 'venstre' | 'høyre' | 'under-høyre' | 'under-venstre';
    /**
     * CSS-stiler for absolutt posisjonering av popover vindu og pil. Brukes av Popover for dynamisk
     * justering av posisjon ved resize og scroll.
     */
    posisjon?: PopoverPosisjonShape;
    /**
     * Variant uten pil på vindu element
     */
    utenPil?: boolean;
    /**
     * onClick callback
     */
    onClick?: (event: React.MouseEvent<HTMLDivElement>) => void;
}

export interface PopoverPosisjonShape {
    left?: number;
    top?: number;
    pilLeft?: number;
}

class PopoverBase extends React.PureComponent<PopoverBaseProps> {
    static defaultProps: Partial<PopoverBaseProps> = {
        orientering: 'under',
        posisjon: {
            left: 0,
            top: 0
        }
    };

    render() {
        const {
            children,
            className,
            apen,
            width,
            height,
            posisjon,
            utenPil,
            ...rest
        } = this.props;

        const stiler = {
            width,
            height,
            left: posisjon!.left,
            top: posisjon!.top
        };

        const pilStiler = (['høyre', 'venstre'].indexOf(this.props.orientering!) === -1)
            ? { left: this.props.posisjon!['pilLeft'] }
            : undefined ;

        return (
            <div
                className={cls(this.props)}
                style={stiler}
                {...rest}
            >
                <div className="popover__content-inner">
                    {children}
                </div>
                {
                    !this.props.utenPil &&
                    <span
                        className="popover__pil"
                        style={pilStiler}
                    />
                }
            </div>
        );
    }
}

export default PopoverBase;
