// nav-frontend-paneler/src/index.tsx
import * as PT from 'prop-types';
import * as React from 'react';
import * as classNames from 'classnames';
import 'nav-frontend-paneler-style';

const cls = (className, border, tabPanel) => classNames('panel', className, {
    'panel--border': border,
    'panel--tabPanel': tabPanel
});

export interface PanelBaseProps {
    /**
     * Innhold
     */
    children: React.ReactNode | React.ReactChild | React.ReactChildren;
    /**
     * Egendefinert klassenavn
     */
    className?: string;
    /**
     * Hvis komponenten skal brukes på hvit bakgrunn kan denne brukes for å gi den en border
     */
    border?: boolean;
    /**
     * Hvis komponenten skal brukes i kombinasjon med Tabs kan denne brukes for å gi den en passende border
     */
    tabPanel?: boolean;
}

class PanelBase extends React.Component<PanelBaseProps> {
    static defaultProps: Partial<PanelBaseProps> = {
        border: false,
        tabPanel: false
    };
    render() {
        const { children, className, border, tabPanel, ...props } = this.props;
        return (
            <div className={cls(className, border, tabPanel)} {...props}>{children}</div>
        );
    }
}

(PanelBase as React.ComponentClass).propTypes = {
    className: PT.string,
    children: PT.node.isRequired,
    border: PT.bool,
    tabPanel: PT.bool
};

(PanelBase as React.ComponentClass).defaultProps = {
    className: undefined,
    border: false,
    tabPanel: false
};

export default PanelBase;

export const Panel = (props) => <PanelBase children={props.children} {...props}/>;
