// nav-frontend-toggle/src/toggleKnappBase.tsx
import * as React from 'react';
import { HTMLAttributes } from 'react';
import * as cn from 'classnames';

import { omit, guid } from 'nav-frontend-js-utils';
import 'nav-frontend-toggle-style';

const knappCls = (props) => cn('knapp', 'toggleKnapp', {
    'toggleKnapp--pressed': props.pressed
});

export interface ToggleKnappBaseProps extends HTMLAttributes<HTMLButtonElement> {
    /**
     * Pressed state
     */
    pressed?: boolean;
    /**
     * onClick handler
     */
    onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

class ToggleKnappBase extends React.Component<ToggleKnappBaseProps> {
    static defaultProps: ToggleKnappBaseProps = {
        pressed: false
    };
    render() {
        const renderProps = omit(this.props, 'children', 'pressed');
        const knappId = guid();
        return (
            <button
                id={knappId}
                className={knappCls(this.props)}
                {...renderProps}
            >
                {this.props.children}
            </button>
        );
    }
}

export default ToggleKnappBase;
