// nav-frontend-toggle/src/toggle-knapp.js
import React, { Component, PropTypes as PT } from 'react';
import { guid } from 'nav-frontend-js-utils';
import classNames from 'classnames';

class ToggleKnapp extends Component { // eslint-disable-line react/prefer-stateless-function
    render() {
        const { children, name, defaultChecked, value, onChange, checked, className } = this.props;
        const knappId = guid();
        const cls = (classes) => classNames('toggle__label', classes);
        return (
            <div>
                <input
                    type="radio"
                    className="toggle__input sr-only"
                    defaultChecked={defaultChecked}
                    onChange={onChange}
                    name={name}
                    id={knappId}
                    value={value}
                    checked={checked}
                />
                <label className={cls(className)} htmlFor={knappId}>{children}</label>
            </div>
        );
    }
}

ToggleKnapp.defaultProps = {
    defaultChecked: undefined,
    checked: undefined,
    className: undefined
};

ToggleKnapp.propTypes = {
    /**
     * Innhold i label
     */
    children: PT.node.isRequired,
    /**
     * Navnet som brukes på radio button gruppen
     */
    name: PT.string.isRequired,
    /**
     * Funksjon som håndterer endring av valgt knapp.
     * Funksjonen tar event. Iden til det klikkede elementet ligger på e.target.value
     * onChange kalles hver gang en knapp blir trykket.
     */
    onChange: PT.func.isRequired,
    /**
     * Mulighet til å oppgi iden til en knapp som vil være satt som default
     */
    defaultChecked: PT.string,
    /**
     * Value som brukes på alle ToggleKnapper i gruppen
     */
    value: PT.string.isRequired,
    /**
     * Marker en knapp som valgt
     */
    checked: PT.bool,
    /**
     * Classname
     */
    className: PT.string
};

export default ToggleKnapp;
