// nav-frontend-toggle/src/toggle.js
import React, { Component, PropTypes as PT } from 'react';
import classNames from 'classnames';
import 'nav-frontend-toggle-style'; // eslint-disable-line import/extensions


const cls = (className) => classNames('toggle', className);
/**
 * Toggleknapp
 */
class Toggle extends Component { // eslint-disable-line react/prefer-stateless-function
    render() {
        const { className, name, buttons, onChange, defaultValgt, valgtKnapp } = this.props;

        const isChecked = (id) => (valgtKnapp ? valgtKnapp === id : undefined);

        return (
            <div className={cls(className)}>
                { buttons.map((button) => (
                    <div>
                        <input
                            onChange={onChange}
                            type="radio"
                            name={name}
                            className="toggle__input toggleknapp"
                            id={button.id}
                            value={button.id}
                            defaultChecked={button.id === defaultValgt}
                            checked={isChecked(button.id)}
                        />
                        <label className="toggle__label" htmlFor={button.id}>{button.label}</label>
                    </div>
                )) }
            </div>
        );
    }
}

Toggle.defaultProps = {
    /**
     * Default classname
     */
    className: '',
    /**
     * Default defaultValgt
     */
    defaultValgt: undefined,
    /**
     * Default valgtKnapp
     */
    valgtKnapp: undefined
};

Toggle.propTypes = {
    /**
     * className
     */
    className: PT.string,
    /**
     * Navnet som brukes på radio button gruppen
     */
    name: PT.string.isRequired,
    /**
     * Funksjon som håndterer endring av valgt knapp.
     * Funksjonen tar event. Iden til det klikkede elementet ligger på e.target.value
     * onChange kalles hver gang en knapp blir trykket.
     */
    onChange: PT.func.isRequired,
    /**
     * En liste med knappene som skal rendres.
     * Innholdet i hver knapp er en React node. Dette betyr at endring av ikoner må håndteres av bruker.
     */
    buttons: PT.arrayOf(PT.shape({
        label: PT.node.isRequired,
        id: PT.string.isRequired
    })).isRequired,
    /**
     * Mulighet til å oppgi iden til en knapp som vil være satt som default
     */
    defaultValgt: PT.string,
    /**
     * Mulighet til å sette valgt knapp fra utsiden.
     */
    valgtKnapp: PT.string
};

export default Toggle;
