// nav-frontend-toggle/src/toggleKnapp.tsx
import * as React from 'react';
import * as cn from 'classnames';

import { omit } from 'nav-frontend-js-utils';
import ToggleKnappBase, { ToggleKnappBaseProps } from './toggleKnappBase';

export interface ToggleKnappState {
    pressed: boolean
}

class ToggleKnapp extends React.Component<ToggleKnappBaseProps, ToggleKnappState> {
    static defaultProps: ToggleKnappBaseProps = {
        pressed: false
    };

    constructor(props){
        super(props);
        this.state = {
            pressed: props.pressed
        };
    }

    handleClick = (e) => {
        this.setState({
            pressed: !this.state.pressed
        });

        if (this.props.onClick) this.props.onClick(e);
    }

    render(){
        const renderProps = omit(this.props, 'children', 'pressed');
        return (
            <ToggleKnappBase
                onClick={this.handleClick}
                pressed={this.state.pressed}
                {...renderProps}
            >
                {this.props.children}
            </ToggleKnappBase>
        );
    }
}

export default ToggleKnapp;
