// nav-frontend-toggle/src/toggle-knapp-pure.tsx
import * as React from 'react';
import * as PT from 'prop-types';
import { HTMLAttributes } from 'react';
import * as classnames from 'classnames';

import { omit, guid } from 'nav-frontend-js-utils';
import 'nav-frontend-toggle-style';

const knappCls = (props) => classnames('knapp', 'toggleKnapp', {
    'toggleKnapp--pressed': props.pressed,
    'toggleKnapp--kompakt': props.kompakt
});

export interface ToggleKnappPureProps {
    /**
     * Pressed state
     */
    pressed?: boolean;
    /**
     * Kompakt versjon
     */
    kompakt?: boolean;
    /**
     * onClick callback
     */
    onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

class ToggleKnappPure extends React.PureComponent<ToggleKnappPureProps> {
    static defaultProps: ToggleKnappPureProps = {
        pressed: false
    };

    render() {
        const renderProps = omit(this.props, 'children', 'pressed', 'kompakt');
        const knappId = guid();
        return (
            <button
                type="button"
                id={knappId}
                className={knappCls(this.props)}
                aria-pressed={this.props.pressed}
                {...renderProps}
            >
                {this.props.children}
            </button>
        );
    }
}

export const ToggleKnappPurePropsShape = PT.shape({
    /**
     * Pressed state
     */
    pressed: PT.bool,
    /**
     * Kompakt versjon
     */
    kompakt: PT.bool,
    /**
     * Custom onChange handler
     */
    onChange: PT.func
});

(ToggleKnappPure as React.ComponentClass).propTypes = ToggleKnappPurePropsShape;

export default ToggleKnappPure;
