// nav-frontend-toggle/src/toggle-knapp-pure.tsx
import * as React from 'react';
import { HTMLAttributes } from 'react';
import * as cn from 'classnames';

import { omit, guid } from 'nav-frontend-js-utils';
import 'nav-frontend-toggle-style';

const knappCls = (props) => cn('knapp', 'toggleKnapp', {
    'toggleKnapp--pressed': props.pressed
});

export interface ToggleKnappPureProps extends HTMLAttributes<HTMLButtonElement> {
    /**
     * Pressed state
     */
    pressed?: boolean;
    /**
     * onClick handler
     */
    onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

class ToggleKnappPure extends React.PureComponent<ToggleKnappPureProps> {
    static defaultProps: ToggleKnappPureProps = {
        pressed: false
    };
    render() {
        const renderProps = omit(this.props, 'children', 'pressed');
        const knappId = guid();
        return (
            <button
                type="button"
                id={knappId}
                className={knappCls(this.props)}
                aria-pressed={this.props.pressed}
                {...renderProps}
            >
                {this.props.children}
            </button>
        );
    }
}

export default ToggleKnappPure;
