// nav-frontend-etiketter/src/index.tsx
import * as PT from 'prop-types';
import * as React from 'react';
import * as classNames from 'classnames';
import TypografiBase from 'nav-frontend-typografi';
import 'nav-frontend-etiketter-style';

const cls = (type, className) => classNames('etikett', className, {
    'etikett--advarsel': type === 'advarsel',
    'etikett--fokus': type === 'fokus',
    'etikett--suksess': type === 'suksess',
    'etikett--info': type === 'info'
});

export interface EtikettProps extends React.DetailedHTMLProps<React.HTMLAttributes<HTMLDivElement>, HTMLDivElement> {
    /**
     * Må være en gyldig 'type' prop på Typografi-komponenten, se 'nav-frontend-typografi'
     */
    typo?: string;
}
/**
 * Etiketter finnes i fire predefinerte farger,
 * om det er behov anbefales å lage en egen klasse hvor man kan legge til flere farger
 */
export interface EtikettBaseProps extends  EtikettProps {
    /**
     * Predefinerte farger
     */
    type: 'suksess' | 'info' | 'advarsel' | 'fokus';
}

function EtikettBase(props: EtikettBaseProps): JSX.Element {
    const { type, className, typo = 'normaltekst', children, ...rest } = props;
    return (
        <div className={cls(type, className)} {...rest}>
            <TypografiBase tag="span" type={typo}>{children}</TypografiBase>
        </div>
    )
}

(EtikettBase as React.StatelessComponent).propTypes = {
    /**
     * Typografi som skal brukes i etiketten. Støtter samme typer som `TypografiBase`. Påvirker ikke tag...
     */
    typo: PT.string,
    children: PT.node.isRequired,
    /**
     * Predefinerte farger
     */
    type: PT.oneOf(['suksess', 'info', 'fokus', 'advarsel']).isRequired,
    className: PT.string
};

(EtikettBase as React.StatelessComponent).defaultProps = {
    typo: 'normaltekst',
    className: undefined
};

export default EtikettBase;

export { default as EtikettAdvarsel } from './etikettadvarsel';
export { default as EtikettFokus } from './etikettfokus';
export { default as EtikettSuksess } from './etikettsuksess';
export { default as EtikettInfo } from './etikettinfo';
