// nav-frontend-skjema/src/bekreft-checkboks-panel.tsx
import * as React from 'react';
import * as classNames from 'classnames';
import * as PT from 'prop-types';
import { guid } from "nav-frontend-js-utils";

export interface Props {
    onChange: (event: React.SyntheticEvent<EventTarget>) => void;
    checked: boolean;
    label: string;
    children?: React.ReactNode | React.ReactChildren;
    inputProps?: React.InputHTMLAttributes<HTMLInputElement>;
    className?: string;
}

export interface State {
    hasFocus: boolean;
    isHover: boolean;
}

export class BekreftCheckboksPanel extends React.Component<Props, State> {
    private childrenId: string;

    constructor(props: Props) {
        super(props);
        this.state = {hasFocus: false, isHover: false};
        if (this.props.children) {
            this.childrenId = guid();
        }
    }

    toggleOutline() {
        return () => this.setState({ hasFocus: !this.state.hasFocus });
    }

    setHover(isHover: boolean) {
        return () => this.setState({ isHover });
    }

    onKeyDown(event: React.KeyboardEvent<EventTarget>) {
        if (event.key === 'Enter') {
            this.props.onChange(event);
        }
    }

    renderChildren() {
        const children = this.props.children;
        return children ? <div className="bekreftCheckboksPanel-innhold" id={this.childrenId}>{children}</div> : null;
    }

    render() {
        const {checked, onChange, label, inputProps, className} = this.props;
        const {hasFocus, isHover} = this.state;

        const cls = classNames('inputPanel bekreftCheckboksPanel', className, {
            'inputPanel--checked': checked,
            'inputPanel--focused': hasFocus,
            'inputPanel--hover': isHover,
        });

        return (
            <div className={cls}>
                {this.renderChildren()}
                <label
                    className="inputPanel__label"
                    onMouseEnter={this.setHover(true)}
                    onMouseLeave={this.setHover(false)}
                    onKeyDown={(event) => this.onKeyDown(event)}
                >
                    <input
                        {...inputProps}
                        className="inputPanel__field"
                        type="checkbox"
                        checked={checked}
                        onFocus={this.toggleOutline()}
                        onBlur={this.toggleOutline()}
                        onChange={onChange}
                        aria-labelledby={this.childrenId}
                    />
                    <span className="inputPanel__label">{label}</span>
                </label>
            </div>
        );
    }
}

(BekreftCheckboksPanel as React.ComponentClass).propTypes = {
    children: PT.node,
    className: PT.string,
    checked: PT.bool.isRequired,
    label: PT.string.isRequired,
    onChange: PT.func.isRequired,
};

export default BekreftCheckboksPanel;
