// nav-frontend-skjema/src/checkboks-panel-gruppe.tsx
import * as React from 'react';
import * as PT from 'prop-types';
import * as classNames from 'classnames';
import { guid } from 'nav-frontend-js-utils';
import 'nav-frontend-skjema-style';
import { SkjemaGruppe, Fieldset } from './';
import { SkjemaelementFeil, skjemaelementFeilmeldingShape } from './skjemaelement-feilmelding';
import { CheckboksPanel, CheckboksProps } from './checkboks-panel';

export interface CheckboksPanelGruppeProps {
    /**
     * Array av checkbokser, se 'checkboks-panel.tsx'
     */
    checkboxes: CheckboksProps[];
    /**
     * Overskrift
     */
    legend: string;
    /**
     * Callback-funksjon som blir kalt straks noen av checkboksene endrer state
     */
    onChange: (event: React.SyntheticEvent<EventTarget>, value?: string) => void;
    /**
     * Objekt som beskriver skjema-feil, se 'skjemaelement-feilmelding.tsx'
     */
    feil?: SkjemaelementFeil;
    /**
     * I enkelte skjemaer vil man ha mer fleksibel barn håndtering. Fieldset har
     * strenge restriksjoner og dette valget gir mulighet til å rendre innholdet 
     * i en div.
     */
    renderAsDiv?: boolean;
}

class CheckboksPanelGruppe extends React.Component<CheckboksPanelGruppeProps> {
    render() {
        const { checkboxes, legend, feil, onChange, renderAsDiv } = this.props;
        return (
            <SkjemaGruppe className="inputPanelGruppe" feil={feil}>
                {renderAsDiv ? 
                    <div role="group">
                        {checkboxes.map((checkbox: CheckboksProps) => (
                            <CheckboksPanel
                                key={`${checkbox.id}${checkbox.label}`}
                                onChange={(event: React.SyntheticEvent<EventTarget>) => onChange(event, checkbox.value)}
                                {...checkbox}
                            />
                        ))}
                    </div>
                :
                    <Fieldset legend={legend}>
                        {checkboxes.map((checkbox: CheckboksProps) => (
                            <CheckboksPanel
                                key={`${checkbox.id}${checkbox.label}`}
                                onChange={(event: React.SyntheticEvent<EventTarget>) => onChange(event, checkbox.value)}
                                {...checkbox}
                            />
                        ))}
                    </Fieldset>
                }
            </SkjemaGruppe>
        );
    }
}

(CheckboksPanelGruppe as React.ComponentClass).propTypes = {
    checkboxes: PT.array.isRequired,
    legend: PT.string.isRequired,
    onChange: PT.func.isRequired,
    feil: skjemaelementFeilmeldingShape
};

(CheckboksPanelGruppe as React.ComponentClass).defaultProps = {
    feil: undefined
};

export default CheckboksPanelGruppe;
