// nav-frontend-skjema/src/fancy-feiloppsummering.tsx
import * as React from 'react';
import * as classnames from 'classnames';
import Lenke from 'nav-frontend-lenker';
import Ikon from 'nav-frontend-ikoner-assets';
import { Undertittel } from 'nav-frontend-typografi';
import { SkjemaelementFeil } from './skjemaelement-feilmelding';
import 'nav-frontend-skjema-style';

const cls = (props) => classnames('fancy-feiloppsummering', props.className, {
    'fancy-feiloppsummering--valid': props.feil.every((feil) => feil.valid),
    'fancy-feiloppsummering--delvis': props.feil.some((feil) => feil.valid) && props.feil.some((feil) => !!feil.valid),
    'fancy-feiloppsummering--invalid': props.feil.every((feil) => !feil.valid)
});

export interface FeiloppsummeringFeil extends SkjemaelementFeil {
    key: string;
    valid: boolean;
}

export interface FancyFeiloppsummeringProps extends React.HTMLAttributes<HTMLDivElement> {
    /**
     * Egendefinert klassenavn.
     */
    className?: string;
    /**
     * Referanse til intern div
     */
    innerRef?: React.RefObject<HTMLDivElement>
    /**
     * Liste over feil
     */
    feil: FeiloppsummeringFeil[];
}

class FancyFeiloppsummering extends React.Component<FancyFeiloppsummeringProps> {

    renderListItems(){
        return this.props.feil.map((feil) => (
            <li key={feil.key} className="fancy-feiloppsummering__listitem">
                <Ikon size="24" kind={(feil.valid) ? 'ok-sirkel-fyll' : 'feil-sirkel-fyll' } />
                <Lenke href={`#${feil.key}`}>
                    {feil.feilmelding}
                </Lenke>
            </li>
        ));
    }

    render(){
        const { children, className, innerRef, feil, ...rest } = this.props;
        return (
            <div
                ref={innerRef}
                tabIndex={0}
                role="region"
                className={cls(this.props)}
                {...rest}
            >
                {children}
                <ul className="fancy-feiloppsummering__list">
                    {this.renderListItems()}
                </ul>
            </div>
        );
    }
}

export default FancyFeiloppsummering;
