// nav-frontend-skjema/src/input-fodselsnr.tsx
import * as React from 'react';
import Input from './input';

interface Props {
    onChange: (value: string) => void;
    submitted: boolean;
    label: string;
    error: string | null;
    bredde?: 'fullbredde' | 'XXL' | 'XL' | 'L' | 'M' | 'S' | 'XS' | 'XXS';
    value: string;
}

class InputFodselsnr extends React.Component<Props> {
    state = { blur: false, value: this.props.value || '' };
    setBlur (blur: boolean) {
        this.setState({ blur });
    }
    render () {
        const updateValue = (e) => {
            if (e.currentTarget.validity.valid) {
                this.setState({ ...this.state, value: e.currentTarget.value.slice(0, 11) });
            }
        };
        const { label, error, submitted, bredde } = this.props;

        return (
            <Input
                bredde={bredde}
                label={label}
                value={this.state.value}
                onChange={(event) => this.props.onChange(event.currentTarget.value)}
                onInput={updateValue}
                onBlur={() => this.setBlur(true)}
                type="text"
                pattern="[0-9]*"
                feil={error && (blur || submitted) ? { feilmelding: error } : undefined}
            />
        );
    }
}

export default InputFodselsnr;
