// nav-frontend-skjema/src/input.js
/* eslint-disable react/prefer-stateless-function  */
import React, { Component, PropTypes as PT } from 'react';
import { guid } from 'nav-frontend-js-utils';
import classNames from 'classnames';

/**
 * Standard inputfelt med mulighet for å sette bredde
 */
class Input extends Component {
    render() {
        const { label, disabled, bredde, feil, id, name } = this.props;

        const inputId = id || name || guid();
        const inputClass = (width) => classNames('skjemaelement__input', `input--${width}`);
        const skjemaElementClass = (harFeil) => classNames('skjemaelement', { harFeil });
        return (
            <div className={skjemaElementClass(feil)}>
                <label className="skjemaelement__label" htmlFor={inputId}>{label}</label>
                <input disabled={disabled} type="text" className={inputClass(bredde)} id={inputId} />
                {feil ?
                    <div role="alert" aria-live="assertive" className="skjemaelement__feilmelding">
                        {feil.feilmelding}
                    </div>
                    : null
                }
            </div>
        );
    }
}

Input.propTypes = {
    /**
     * Label for tekstfeltet
     */
    label: PT.node.isRequired,
    /**
     * Hvis skjemaet har feil sender man inn et objekt med en feilmelding
     */
    feil: PT.shape({
        feilmelding: PT.node.isRequired
    }),
    /**
     * Bredden til inputfeltet
     */
    bredde: PT.oneOf['fullbredde', 'xxl', 'xl', 'l', 'm', 's', 'xs', 'xxs'], // eslint-disable-line no-sequences
    /**
     * Mulighet for å sette inputfeltet til disabled
     */
    disabled: PT.bool,
    /**
     * Id for inputfelt, hvis id ikke er satt brukes name eller en tilfeldig guid
     */
    id: PT.string,
    /**
     * Name for inputfelt, hvis name ikke er satt brukes id eller en tilfeldig guid
     */
    name: PT.string
};

Input.defaultProps = {
    bredde: 'fullbredde',
    disabled: false,
    feil: undefined,
    id: undefined,
    name: undefined
};

export default Input;
