// nav-frontend-skjema/src/radio.js
import PT from 'prop-types';
/* eslint-disable react/prefer-stateless-function  */
import React, { Component } from 'react';
import { guid } from 'nav-frontend-js-utils';
import 'nav-frontend-skjema-style'; // eslint-disable-line import/extensions
import classNames from 'classnames';

const cls = (className) => classNames('skjemaelement skjemaelement--horisontal', className);

/**
 * Radio
 */
class Radio extends Component {

    render() {
        const { className, disabled, name, label, id, radioRef, ...other } = this.props;
        const inputId = id || guid();
        return (
            <div className={cls(className)}>
                <input
                    type="radio"
                    name={name}
                    className="skjemaelement__input radioknapp"
                    id={inputId}
                    ref={radioRef}
                    disabled={disabled}
                    {...other}
                />
                <label className="skjemaelement__label" htmlFor={inputId}>{label}</label>
            </div>
        );
    }
}

Radio.propTypes = {
    /**
     * ClassName
     */
    className: PT.string,
    /**
     * Label for radiobutton
     */
    label: PT.node.isRequired,
    /**
     * Id for radiobutton, hvis id ikke er satt brukes en tilfeldig guid
     */
    id: PT.string,
    /**
     * Name for radiobutton gruppe
     */
    name: PT.string.isRequired,
    /**
     * Referanse til selve radioknappen. Brukes for eksempel til å sette fokus
     */
    radioRef: PT.func,
    disabled: PT.bool
};

Radio.defaultProps = {
    className: undefined,
    id: undefined,
    radioRef: undefined,
    disabled: false
};

export default Radio;
