// nav-frontend-skjema/src/skjemaelement-feilmelding.tsx
import * as React from 'react';
import * as PT from 'prop-types';

export interface SkjemaelementFeil {
    feilmelding: string | React.ReactNode;
}

export interface SkjemaelementFeilmeldingProps {
    feil?: SkjemaelementFeil;
}

class SkjemaelementFeilmelding extends React.Component<SkjemaelementFeilmeldingProps> {

    render() {
        const { feil } = this.props;
        return (
            <div role="alert" aria-live="assertive">
                {feil && <div className="skjemaelement__feilmelding">{feil.feilmelding}</div>}
            </div>
        );
    }
}

export const skjemaelementFeilmeldingShape = PT.shape({
    feilmelding: PT.oneOfType([
        PT.string,
        PT.node // For å slippe igjennom FormattedMessage fra react-intl
    ]).isRequired
});

(SkjemaelementFeilmelding as React.ComponentClass).propTypes = {
    /**
     * Hvis skjemaet har feil sender man inn et objekt med en feilmelding
     */
    feil: skjemaelementFeilmeldingShape
};

(SkjemaelementFeilmelding as React.ComponentClass).defaultProps = {
    feil: undefined
};

export default SkjemaelementFeilmelding;
