// nav-frontend-skjema/src/textarea.js
import React, { Component, PropTypes as PT } from 'react';
import { guid, autobind, omit } from 'nav-frontend-js-utils';
import 'nav-frontend-skjema-style'; // eslint-disable-line import/extensions
import classNames from 'classnames';

/**
 * Selvekspanderende tekstområde med teller
 */
class Textarea extends Component {

    constructor(props) {
        super(props);
        const initialValue = props.defaultValue || props.value || '';
        this.state = { antall: initialValue.length };
        autobind(this);
    }

    componentDidMount() {
        window.requestAnimationFrame = window.requestAnimationFrame || ((fn) => setTimeout(fn, 0));
        window.requestAnimationFrame(() => {
            if (this.tekstomrade) {
                this.tekstomrade.style.height = `${Math.max(this.tekstomrade.scrollHeight, 30) + 30}px`;
            }
        });
    }

    handleChange(e) {
        // eslint-disable-next-line no-param-reassign
        e.target.style.height = `${30}px`;
        // eslint-disable-next-line no-param-reassign
        e.target.style.height = `${Math.max(this.tekstomrade.scrollHeight, 30) + 30}px`;

        this.setState({
            antall: this.tekstomrade.value.length
        });

        if (this.props.onChange !== undefined) {
            this.props.onChange(e);
        }
    }

    render() {
        const { label, maxLength, textareaClass, id, name, ...other } = this.props;

        const domProps = omit(other, 'onChange');
        const textareaId = id || name || guid();
        const cls = (className) => classNames(className, 'skjemaelement__input textarea--medMeta');

        return (
            <div className="skjemaelement">
                <label className="skjemaelement__label" htmlFor={id}>
                    {label}
                </label>
                <textarea
                    ref={(textarea) => { this.tekstomrade = textarea; }}
                    onChange={this.handleChange}
                    className={cls(textareaClass)}
                    type="text"
                    id={textareaId}
                    style={{ height: '30px' }}
                    maxLength={maxLength}
                    {...domProps}
                />
                <p className="textarea--medMeta__teller">
                    <span>Du har {maxLength - this.state.antall} tegn igjen</span>
                </p>
            </div>
        );
    }
}

Textarea.propTypes = {
    /**
     * Ledetekst for tekstområdet
     */
    label: PT.node.isRequired,
    /**
     * Maks antal tegn som kan skrives inn i tekstområdet
     */
    maxLength: PT.number,
    /**
     * Teksten som er skrevet inn i tekstområdet.
     */
    value: PT.string,
    /**
     * Verdi som er preutfyllt i tekstområdet.
     */
    defaultValue: PT.string,
    /**
     * Klassenavn for tekstomnrådet
     */
    textareaClass: PT.string,
    /**
     * Id for tekstområdet, settes til name eller random guid hvis prop ikke er satt
     */
    id: PT.string,
    /**
     * Navn for tekstområdet, settes til id eller random guid hvis prop ikke er satt
     */
    name: PT.string,
    /**
     * Optional onChange handler
     */
    onChange: PT.func
};

Textarea.defaultProps = {
    maxLength: 2000,
    value: undefined,
    defaultValue: undefined,
    textareaClass: '',
    id: undefined,
    name: undefined,
    onChange: undefined
};


export default Textarea;
