// nav-frontend-popover/_popover-orientering.example.js
import React, { Component } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Knapp } from 'nav-frontend-knapper';
// eslint-disable-next-line
import Popover from './';

export default class PopoverOrienteringEksempel extends Component {

    constructor(props){
        super(props);
        this.state = {};
    }

    togglePopover = (orientering, anker) => {
        window.clearTimeout(this.timer);
        const newState = Object.assign({}, this.state);
        
        if (orientering === newState.orientering) {
            newState['anker'] = (newState['anker']) ? undefined : anker ;
        } else {
            newState['anker'] = anker;
        }

        newState['orientering'] = orientering;
        this.setState(newState);
    }

    render() {
        return (
            <div>
                <Knapp onClick={(e) => this.togglePopover('over', e.currentTarget)}>Over</Knapp>&nbsp;
                <Knapp onClick={(e) => this.togglePopover('under', e.currentTarget)}>Under</Knapp>&nbsp;
                <Knapp onClick={(e) => this.togglePopover('høyre', e.currentTarget)}>Høyre</Knapp>&nbsp;
                <Knapp onClick={(e) => this.togglePopover('venstre', e.currentTarget)}>Venstre</Knapp><br/><br/>
                <Knapp onClick={(e) => this.togglePopover('over-venstre', e.currentTarget)}>Over venstre</Knapp>&nbsp;
                <Knapp onClick={(e) => this.togglePopover('over-høyre', e.currentTarget)}>Over høyre</Knapp><br/><br/>
                <Knapp onClick={(e) => this.togglePopover('under-venstre', e.currentTarget)}>Under venstre</Knapp>&nbsp;
                <Knapp onClick={(e) => this.togglePopover('under-høyre', e.currentTarget)}>Under høyre</Knapp>

                <Popover
                    ankerEl={this.state.anker}
                    onClose={() => this.timer = window.setTimeout(() => this.setState({anker: undefined}), 100)}
                    scrollContainerSelector="article.mainContent"
                    orientering={this.state.orientering}
                >
                    <p style={{margin:'1rem', maxWidth:300}}>
                        Dette er en popover som bruker <code className="inline">orientering</code>-propen til å posisjonere seg i forhold til ankeret sitt.
                    </p>
                </Popover>
            </div>
        );
    }
}
