// nav-frontend-tabs/src/tabs.tsx
import * as React from 'react';
import { guid, omit } from 'nav-frontend-js-utils';

import Tab, {TabProps} from './tab';
import TabsPure, {TabsPureProps} from './tabs-pure';

import 'nav-frontend-tabs-style';

export interface TabsProps extends TabsPureProps {
    onChange: (index: number) => void;
    defaultAktiv?: number;
}

export interface TabsState {
    aktivTab: number;
}

class Tabs extends React.Component<TabsProps, TabsState> {

    static Tab = Tab;

    static defaultProps: Partial<TabsProps> = {
        kompakt: false,
        defaultAktiv: 0
    };

    constructor(props) {
        super(props);
        this.state = {
            aktivTab: props.defaultAktiv
        };
    }

    handleClick = (e, index) => {
        e.preventDefault();
        this.setState({
            aktivTab: index
        });
        this.props.onChange(index);
    }

    renderTabs() {
        if (this.props.children) {
            return React.Children.map(this.props.children, (child, i) => {
                if (React.isValidElement(child)) {
                    return React.cloneElement(child as React.ReactElement<any>, {
                        aktiv: i === this.state.aktivTab,
                        onClick: (!child.props['disabled']) ? (e) => this.handleClick(e, i) : undefined
                    });
                }
                return child;
            });
        }

        return this.props.tabs.map((tab, i) => {
            return (
                <Tab
                    key={guid()}
                    aktiv={(this.state.aktivTab === i)}
                    onClick={(!tab.disabled) ? (e) => this.handleClick(e, i) : undefined}
                    {...tab}
                />
            );
        })
    }

    render() {
        const domProps = omit(this.props, 'tabs', 'onChange', 'defaultAktiv');

        return (
            <TabsPure {...domProps}>
                {this.renderTabs()}
            </TabsPure>
        );
    }
}

export default Tabs;
export { default as TabsPure } from './tabs-pure';
