// nav-frontend-tabs/src/tabs-pure.tsx
import * as React from 'react';
import * as cn from 'classnames';
import { guid, omit } from 'nav-frontend-js-utils';

import Tab, {TabProps} from './tab';

const tabsCls = (props) => cn('nav-frontend-tabs', props.className, {
    'nav-frontend-tabs--kompakt': props.kompakt
});

export interface TabsPureProps {
    tabs: TabProps[];
    kompakt?: boolean;
}

class TabsPure extends React.PureComponent<TabsPureProps> {

    static Tab = Tab;

    static defaultProps: Partial<TabsPureProps> = {
        kompakt: false
    };

    renderTabs() {
        if (this.props.children) return this.props.children;

        return this.props.tabs.map((tab, i) => {
            return (
                <Tab
                    key={guid()}
                    {...tab}
                />
            );
        })
    }

    render () {
        const domProps = omit(this.props,
            'children',
            'className',
            'tabs',
            'kompakt'
        );

        return (
            <div className={tabsCls(this.props)} {...domProps}>
                <ul className="nav-frontend-tabs__tab-list" role="tablist">
                    {this.renderTabs()}
                </ul>
            </div>
        );
    }
}

export default TabsPure;
