// nav-frontend-modal/src/index.js
/* eslint-disable react/prefer-stateless-function  */
import React, { Component, PropTypes as PT } from 'react';
import Modal from 'react-modal';
import classnames from 'classnames';
import 'nav-frontend-modal-style'; // eslint-disable-line import/extensions

const cls = (className) => classnames('modal', className);

/**
 * Modalen bruker `react-modal`, se [github siden](https://github.com/reactjs/react-modal) for mer informasjon.
 *
 */
class ModalWrapper extends Component {
    static setAppElement(element) {
        Modal.setAppElement(element);
    }
    render() {
        const { children, closeButton, ...props } = this.props;
        return (
            <Modal {...props} className={cls(props.className)} overlayClassName="modal__overlay">
                {children}
                { closeButton && <button className="modal__lukkknapp" onClick={props.onRequestClose}>
                    <span className="text-hide">Lukk modal</span>
                </button> }
            </Modal>
        );
    }
}

ModalWrapper.propTypes = {
    /**
     * Bestemmer om modalen selv skal legge til en lukkeknapp
     */
    closeButton: PT.bool,
    /**
     * Bestemmer om modalen skal være synlig
     */
    isOpen: PT.bool.isRequired,
    /**
     * En beskrivelse av formålet med modalen, blir satt som `aria-label`
     */
    contentLabel: PT.string.isRequired,
    /**
     * Innholdet i modalen
     */
    children: PT.node.isRequired,
    /**
     * Funksjon som blir kalt når modalen har blitt åpnet. Kan brukes for å sette fokus på ett element
     */
    onAfterOpen: PT.func,
    /**
     * Funksjon som blir kalt i det modalen ønsker å lukkes
     */
    onRequestClose: PT.func.isRequired,
    /**
     * Tall som beskriver hvor lenge modalen venter før den lukkes
     */
    closeTimeoutMS: PT.number
};

ModalWrapper.defaultProps = {
    closeButton: true,
    closeTimeoutMS: 0,
    onAfterOpen: () => {}
};

export default ModalWrapper;
