// nav-frontend-skjema/fokus-eksempel.js
import React, { Component } from 'react';
import { autobind } from 'nav-frontend-js-utils';
import Input from './src/input';
import Select from './src/select';
import Textarea from './src/textarea';
import Checkbox from './src/checkbox';
import Radio from './src/radio';

class FokusEksempel extends Component {

    constructor(props) {
        super(props);
        autobind(this);
    }

    setFocusInput() {
        this.input.focus();
    }

    setFocusTextarea() {
        this.textarea.focus();
    }

    setFocusSelect() {
        this.select.focus();
    }

    setFocusCheckbox() {
        this.checkbox.focus();
    }

    setFocusRadio() {
        this.radio.focus();
    }

    render() {
        return (
            <div>
                <button onClick={this.setFocusInput}>Klikk for å sette fokus på input</button>
                <Input inputRef={(input) => { this.input = input; }} label="Navn" bredde="xxl" />

                <button onClick={this.setFocusTextarea}>Klikk for å sette fokus på textarea</button>
                <Textarea textareaRef={(textarea) => { this.textarea = textarea; }} label="Adresse" />

                <button onClick={this.setFocusSelect}>Klikk for å sette fokus på select</button>
                <Select selectRef={(select) => { this.select = select; }} label="Velg land" bredde="xs">
                    <option value="norge">Norge</option>
                    <option value="sverige">Sverige</option>
                    <option value="danmark">Danmark</option>
                </Select>

                <button onClick={this.setFocusCheckbox}>Klikk for å sette fokus på checkbox</button>
                <Checkbox checkboxRef={(checkbox) => { this.checkbox = checkbox; }} label="Is" value="is" name="is" />
                <Checkbox label="Brus" value="brus" name="brus" />

                <button onClick={this.setFocusRadio}>Klikk for å sette fokus på radioknapp</button>
                <Radio radioRef={(radio) => { this.radio = radio; }} label="Ja" value="ja" name="svar" />
                <Radio label="Nei" value="nei" name="svar" />
            </div>
        );
    }
}

export default FokusEksempel;
