// nav-frontend-animert-ikon/src/ella.tsx
import * as React from 'react';
import Lottie from 'react-lottie';
import * as PT from "prop-types";

export interface EllaProps {
	hoyde?: number;
	erStoppet?: boolean | null;
	erPauset?: boolean | null;
}

const blunkedamedata: any = {
	"v": "5.1.16",
	"fr": 30,
	"ip": 0,
	"op": 90,
	"w": 100,
	"h": 168,
	"nm": "veileder",
	"ddd": 0,
	"assets": [],
	"layers": [ {
		"ddd": 0,
		"ind": 1,
		"ty": 4,
		"nm": "øyne Outlines",
		"sr": 1,
		"ks": {
			"o": { "a": 0, "k": 100, "ix": 11 },
			"r": { "a": 0, "k": 0, "ix": 10 },
			"p": { "a": 0, "k": [ 50, 59.624, 0 ], "ix": 2 },
			"a": { "a": 0, "k": [ 21.093, 11.389, 0 ], "ix": 1 },
			"s": {
				"a": 1,
				"k": [ {
					"i": { "x": [ 0.833, 0.833, 0.833 ], "y": [ 0.833, 0.833, 0.833 ] },
					"o": { "x": [ 0.167, 0.167, 0.167 ], "y": [ 0.167, 0.167, 0.167 ] },
					"n": [ "0p833_0p833_0p167_0p167", "0p833_0p833_0p167_0p167", "0p833_0p833_0p167_0p167" ],
					"t": 0,
					"s": [ 100, 100, 100 ],
					"e": [ 100, 0, 100 ]
				}, {
					"i": { "x": [ 0.833, 0.833, 0.833 ], "y": [ 0.833, 0.833, 0.833 ] },
					"o": { "x": [ 0.167, 0.167, 0.167 ], "y": [ 0.167, 0.167, 0.167 ] },
					"n": [ "0p833_0p833_0p167_0p167", "0p833_0p833_0p167_0p167", "0p833_0p833_0p167_0p167" ],
					"t": 6,
					"s": [ 100, 0, 100 ],
					"e": [ 100, 100, 100 ]
				}, { "t": 12 } ],
				"ix": 6
			}
		},
		"ao": 0,
		"shapes": [ {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0.085, -0.186 ], [ -2.597, -0.197 ], [ 0.716, 2.289 ], [ 1.391, 0 ] ],
						"o": [ [ -1.023, 2.231 ], [ 2.374, 0.179 ], [ -0.134, -0.434 ], [ -1.39, 0 ] ],
						"v": [ [ 11.171, -2.835 ], [ 13.247, 3.922 ], [ 15.57, -1.692 ], [ 13.259, -4.101 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ind": 1,
				"ty": "sh",
				"ix": 2,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0.134, -0.434 ], [ -2.374, 0.179 ], [ 1.022, 2.231 ], [ 1.391, 0 ] ],
						"o": [ [ -0.715, 2.289 ], [ 2.597, -0.197 ], [ -0.085, -0.186 ], [ -1.39, 0 ] ],
						"v": [ [ -15.571, -1.692 ], [ -13.248, 3.922 ], [ -11.171, -2.835 ], [ -13.26, -4.101 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 2",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "mm",
				"mm": 1,
				"nm": "Merge Paths 1",
				"mn": "ADBE Vector Filter - Merge",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.315303039551, 0.262281111175, 0.238536385929, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 21.094, 9.35 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 1",
			"np": 4,
			"cix": 2,
			"ix": 1,
			"mn": "ADBE Vector Group",
			"hd": false
		} ],
		"ip": 0,
		"op": 900,
		"st": 0,
		"bm": 0
	}, {
		"ddd": 0,
		"ind": 2,
		"ty": 4,
		"nm": "nese Outlines",
		"sr": 1,
		"ks": {
			"o": { "a": 0, "k": 100, "ix": 11 },
			"r": { "a": 0, "k": 0, "ix": 10 },
			"p": { "a": 0, "k": [ 49.999, 68.362, 0 ], "ix": 2 },
			"a": { "a": 0, "k": [ 9.484, 9.823, 0 ], "ix": 1 },
			"s": { "a": 0, "k": [ 100, 100, 100 ], "ix": 6 }
		},
		"ao": 0,
		"shapes": [ {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0.523, -0.081 ], [ -0.069, -0.449 ], [ -0.447, 0.07 ], [ -0.342, -0.42 ], [ 2.211, -1.378 ], [ 0.848, 0.376 ], [ 0.184, -0.415 ], [ -0.414, -0.184 ], [ -1.587, 0.989 ], [ 1.952, 2.403 ], [ 1.209, 0 ] ],
						"o": [ [ -0.448, 0.068 ], [ 0.069, 0.448 ], [ 1.396, -0.215 ], [ 1.288, 1.586 ], [ -1.163, 0.723 ], [ -0.414, -0.184 ], [ -0.182, 0.415 ], [ 1.4, 0.62 ], [ 3.005, -1.872 ], [ -0.601, -0.739 ], [ -0.451, 0 ] ],
						"v": [ [ -1.448, -4.604 ], [ -2.134, -3.667 ], [ -1.199, -2.981 ], [ 1.493, -2.641 ], [ 0.158, 2.002 ], [ -3.452, 2.605 ], [ -4.535, 3.024 ], [ -4.116, 4.108 ], [ 1.025, 3.397 ], [ 2.765, -3.678 ], [ 0.016, -4.728 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.81223042806, 0.724314311906, 0.557397341261, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 9.855, 9.978 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 1",
			"np": 2,
			"cix": 2,
			"ix": 1,
			"mn": "ADBE Vector Group",
			"hd": false
		} ],
		"ip": 0,
		"op": 900,
		"st": 0,
		"bm": 0
	}, {
		"ddd": 0,
		"ind": 3,
		"ty": 4,
		"nm": "munn Outlines",
		"sr": 1,
		"ks": {
			"o": { "a": 0, "k": 100, "ix": 11 },
			"r": { "a": 0, "k": 0, "ix": 10 },
			"p": { "a": 0, "k": [ 50, 81.158, 0 ], "ix": 2 },
			"a": { "a": 0, "k": [ 17.479, 9.387, 0 ], "ix": 1 },
			"s": { "a": 0, "k": [ 100, 100, 100 ], "ix": 6 }
		},
		"ao": 0,
		"shapes": [ {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0.135, -0.309 ], [ 0.273, -0.429 ], [ 0.704, -0.676 ], [ 3.474, 0.107 ], [ 2.132, 1.853 ], [ 0.516, 0.726 ], [ 0.083, 0.169 ], [ 0.303, 0 ], [ 0.116, -0.058 ], [ -0.2, -0.406 ], [ -0.359, -0.505 ], [ -0.885, -0.769 ], [ -3.783, -0.116 ], [ -2.402, 2.312 ], [ -0.528, 0.829 ], [ -0.107, 0.245 ], [ 0.415, 0.181 ], [ 0.11, 0 ] ],
						"o": [ [ -0.073, 0.17 ], [ -0.462, 0.724 ], [ -2.098, 2.019 ], [ -3.39, -0.103 ], [ -0.782, -0.681 ], [ -0.305, -0.429 ], [ -0.142, -0.291 ], [ -0.121, 0 ], [ -0.406, 0.2 ], [ 0.121, 0.245 ], [ 0.588, 0.827 ], [ 2.406, 2.093 ], [ 3.939, 0.121 ], [ 0.804, -0.772 ], [ 0.321, -0.504 ], [ 0.181, -0.415 ], [ -0.106, -0.047 ], [ -0.316, 0 ] ],
						"v": [ [ 10.658, -3.7 ], [ 10.141, -2.777 ], [ 8.396, -0.65 ], [ 0.092, 2.43 ], [ -8.148, -0.676 ], [ -10.092, -2.811 ], [ -10.671, -3.734 ], [ -11.406, -4.194 ], [ -11.768, -4.108 ], [ -12.143, -3.009 ], [ -11.428, -1.858 ], [ -9.223, 0.564 ], [ 0.042, 4.072 ], [ 9.533, 0.535 ], [ 11.525, -1.893 ], [ 12.161, -3.043 ], [ 11.738, -4.124 ], [ 11.411, -4.194 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.81832634421, 0.556886500938, 0.540723134957, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 17.476, 9.444 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 1",
			"np": 2,
			"cix": 2,
			"ix": 1,
			"mn": "ADBE Vector Group",
			"hd": false
		} ],
		"ip": 0,
		"op": 900,
		"st": 0,
		"bm": 0
	}, {
		"ddd": 0,
		"ind": 4,
		"ty": 4,
		"nm": "hode Outlines",
		"sr": 1,
		"ks": {
			"o": { "a": 0, "k": 100, "ix": 11 },
			"r": { "a": 0, "k": 0, "ix": 10 },
			"p": { "a": 0, "k": [ 49.999, 65.149, 0 ], "ix": 2 },
			"a": { "a": 0, "k": [ 42.164, 38.111, 0 ], "ix": 1 },
			"s": { "a": 0, "k": [ 100, 100, 100 ], "ix": 6 }
		},
		"ao": 0,
		"shapes": [ {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 2.468, -5.083 ], [ 0.222, 0.161 ], [ 0.353, 0.001 ], [ 0.004, -1.58 ], [ 0, 0 ], [ -1.525, -0.003 ], [ -0.458, 0.349 ], [ -15.66, -0.027 ], [ -3.735, 17.598 ], [ -0.61, -0.001 ], [ -0.002, 1.578 ], [ 0, 0 ], [ 1.523, 0.003 ], [ 0.309, -0.124 ], [ 1.041, 0.002 ] ],
						"o": [ [ 0, 0 ], [ -0.311, -0.125 ], [ -1.522, -0.002 ], [ 0, 0 ], [ -0.004, 1.58 ], [ 0.61, 0 ], [ 3.685, 17.611 ], [ 15.659, 0.025 ], [ 0.455, 0.349 ], [ 1.524, 0.002 ], [ 0, 0 ], [ 0.002, -1.581 ], [ -0.352, 0 ], [ -35.316, 4.192 ], [ 0, 0 ] ],
						"v": [ [ -32.747, -13.944 ], [ -33.141, -13.128 ], [ -34.136, -13.329 ], [ -36.899, -10.47 ], [ -36.913, 0.466 ], [ -34.161, 3.332 ], [ -32.535, 2.773 ], [ -0.058, 32.848 ], [ 32.51, 2.88 ], [ 34.135, 3.444 ], [ 36.898, 0.587 ], [ 36.915, -10.35 ], [ 34.162, -13.217 ], [ 33.167, -13.019 ], [ -15.509, -32.874 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.915745275161, 0.901643640855, 0.871193739947, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 42.164, 38.124 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 1",
			"np": 2,
			"cix": 2,
			"ix": 1,
			"mn": "ADBE Vector Group",
			"hd": false
		} ],
		"ip": 0,
		"op": 900,
		"st": 0,
		"bm": 0
	}, {
		"ddd": 0,
		"ind": 5,
		"ty": 4,
		"nm": "Kropp Outlines",
		"sr": 1,
		"ks": {
			"o": { "a": 0, "k": 100, "ix": 11 },
			"r": { "a": 0, "k": 0, "ix": 10 },
			"p": { "a": 0, "k": [ 50.001, 125.941, 0 ], "ix": 2 },
			"a": { "a": 0, "k": [ 49.547, 43.04, 0 ], "ix": 1 },
			"s": { "a": 0, "k": [ 100, 100, 100 ], "ix": 6 }
		},
		"ao": 0,
		"shapes": [ {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0 ] ],
						"o": [ [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0 ] ],
						"v": [ [ -0.763, 2.054 ], [ 0.763, 2.054 ], [ 0.763, -2.054 ], [ -0.763, -2.054 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.767070037243, 0.67057315602, 0.782972388174, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 72.726, 40.166 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 1",
			"np": 2,
			"cix": 2,
			"ix": 1,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0, -0.193 ], [ 0, 0 ], [ -0.195, 0 ], [ 0, 0 ], [ 0, 0.195 ], [ 0, 0 ], [ 0.196, 0 ] ],
						"o": [ [ -0.195, 0 ], [ 0, 0 ], [ 0, 0.195 ], [ 0, 0 ], [ 0.196, 0 ], [ 0, 0 ], [ 0, -0.193 ], [ 0, 0 ] ],
						"v": [ [ -1.316, -0.573 ], [ -1.669, -0.221 ], [ -1.669, 0.221 ], [ -1.316, 0.573 ], [ 1.318, 0.573 ], [ 1.67, 0.221 ], [ 1.67, -0.221 ], [ 1.318, -0.573 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.248701626647, 0.016735363007, 0.235442202699, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 72.725, 42.052 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 2",
			"np": 2,
			"cix": 2,
			"ix": 2,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, -1.007 ], [ -0.396, 0 ], [ 0, 0 ], [ -0.017, 0.036 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0.23 ], [ -0.459, 0 ], [ 0, 0 ], [ 0, 0.63 ], [ 0.532, 0 ] ],
						"o": [ [ 0, 1.006 ], [ 0, 0 ], [ 0.04, 0 ], [ 0, 0 ], [ 0, 0 ], [ -0.13, 0 ], [ 0, -0.199 ], [ 0.467, 0 ], [ 0, 0 ], [ 0, -0.638 ], [ -0.644, 0 ] ],
						"v": [ [ -1.201, 0.038 ], [ 0.234, 1.564 ], [ 0.615, 1.564 ], [ 0.706, 1.505 ], [ 1.201, 0.415 ], [ 0.365, 0.415 ], [ -0.001, -0.031 ], [ 0.562, -0.497 ], [ 1.12, 0.106 ], [ 1.201, -0.561 ], [ 0.215, -1.564 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.791419593961, 0.055595973894, 0.075535015031, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 72.116, 50.774 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 3",
			"np": 2,
			"cix": 2,
			"ix": 3,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ -0.041, -0.102 ], [ 0, 0 ], [ -0.064, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0.067, 0 ] ],
						"o": [ [ -0.114, 0 ], [ 0, 0 ], [ 0.024, 0.063 ], [ 0, 0 ], [ 0, 0 ], [ -0.021, -0.066 ], [ 0, 0 ] ],
						"v": [ [ -1.322, -1.564 ], [ -0.819, -1.351 ], [ 0.316, 1.465 ], [ 0.462, 1.564 ], [ 1.436, 1.564 ], [ 0.434, -1.458 ], [ 0.285, -1.564 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.791419593961, 0.055595973894, 0.075535015031, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 75.129, 50.774 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 4",
			"np": 2,
			"cix": 2,
			"ix": 4,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0.012, -0.03 ], [ 0, 0 ], [ 0, 0 ], [ -0.009, 0.03 ], [ 0, 0 ], [ 0.055, 0 ] ],
						"o": [ [ -0.033, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0.033, 0 ], [ 0, 0 ], [ 0.023, -0.052 ], [ 0, 0 ] ],
						"v": [ [ -0.068, -1.564 ], [ -0.142, -1.513 ], [ -1.162, 1.564 ], [ -0.134, 1.564 ], [ -0.064, 1.516 ], [ 1.139, -1.46 ], [ 1.068, -1.564 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.791419593961, 0.055595973894, 0.075535015031, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 76.906, 50.774 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 5",
			"np": 2,
			"cix": 2,
			"ix": 5,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0.059 ], [ 0, 0 ], [ 0.055, 0 ] ],
						"o": [ [ 0, 0 ], [ 0, 0 ], [ 0.061, 0 ], [ 0, 0 ], [ 0.023, -0.049 ], [ 0, 0 ] ],
						"v": [ [ 0.213, -0.575 ], [ -0.57, 0.575 ], [ 0.059, 0.575 ], [ 0.169, 0.464 ], [ 0.547, -0.469 ], [ 0.478, -0.575 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.791419593961, 0.055595973894, 0.075535015031, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 70.921, 51.763 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 6",
			"np": 2,
			"cix": 2,
			"ix": 6,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0, -0.048 ], [ 0, 0 ], [ -0.049, 0 ], [ 0, 0 ], [ 0, 0.062 ], [ 0, 0 ], [ 0.059, 0 ], [ 0, 0 ], [ 0, 0.04 ], [ 0, 0 ], [ 0.047, 0 ] ],
						"o": [ [ -0.049, 0 ], [ 0, 0 ], [ 0, 0.048 ], [ 0, 0 ], [ 0.064, 0 ], [ 0, 0 ], [ 0.024, -0.05 ], [ 0, 0 ], [ -0.044, 0 ], [ 0, 0 ], [ 0, -0.048 ], [ 0, 0 ] ],
						"v": [ [ -0.696, -1.564 ], [ -0.783, -1.478 ], [ -0.783, 1.478 ], [ -0.696, 1.564 ], [ 0.271, 1.564 ], [ 0.381, 1.452 ], [ 0.758, 0.521 ], [ 0.688, 0.415 ], [ 0.424, 0.415 ], [ 0.351, 0.341 ], [ 0.351, -1.478 ], [ 0.26, -1.564 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.791419593961, 0.055595973894, 0.075535015031, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 74.021, 50.774 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 7",
			"np": 2,
			"cix": 2,
			"ix": 7,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0, -0.062 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, -0.172 ], [ 0, 0 ], [ -0.049, 0 ], [ 0, 0 ], [ 0, 0.048 ], [ 0, 0 ], [ 0.05, 0 ] ],
						"o": [ [ -0.063, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0.17, 0 ], [ 0, 0 ], [ 0, 0.048 ], [ 0, 0 ], [ 0.05, 0 ], [ 0, 0 ], [ 0, -0.048 ], [ 0, 0 ] ],
						"v": [ [ -0.477, -1.564 ], [ -0.589, -1.452 ], [ -0.969, -0.522 ], [ -1.1, -0.415 ], [ -0.341, -0.415 ], [ -0.033, -0.111 ], [ -0.033, 1.478 ], [ 0.054, 1.564 ], [ 1.008, 1.564 ], [ 1.1, 1.478 ], [ 1.1, -1.478 ], [ 1.008, -1.564 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.791419593961, 0.055595973894, 0.075535015031, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 69.956, 50.774 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 8",
			"np": 2,
			"cix": 2,
			"ix": 8,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0, -0.062 ], [ 0, 0 ], [ -0.052, 0 ], [ 0, 0 ], [ 0, -0.045 ], [ 0, 0 ], [ -0.047, 0 ], [ 0, 0 ], [ 0, 0.048 ], [ 0, 0 ], [ 0.05, 0 ] ],
						"o": [ [ -0.064, 0 ], [ 0, 0 ], [ -0.023, 0.048 ], [ 0, 0 ], [ 0.041, 0 ], [ 0, 0 ], [ 0, 0.048 ], [ 0, 0 ], [ 0.05, 0 ], [ 0, 0 ], [ 0, -0.048 ], [ 0, 0 ] ],
						"v": [ [ -0.269, -1.564 ], [ -0.38, -1.452 ], [ -0.758, -0.522 ], [ -0.693, -0.415 ], [ -0.421, -0.415 ], [ -0.344, -0.341 ], [ -0.344, 1.478 ], [ -0.26, 1.564 ], [ 0.693, 1.564 ], [ 0.781, 1.478 ], [ 0.781, -1.478 ], [ 0.693, -1.564 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.791419593961, 0.055595973894, 0.075535015031, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 68.199, 50.774 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 9",
			"np": 2,
			"cix": 2,
			"ix": 9,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0 ] ],
						"o": [ [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0 ] ],
						"v": [ [ 0.487, -1.58 ], [ -0.805, 1.58 ], [ -0.487, 1.58 ], [ 0.805, -1.58 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.999999940162, 0.999999940162, 1, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 82.424, 50.772 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 10",
			"np": 2,
			"cix": 2,
			"ix": 10,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0 ] ],
						"o": [ [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0 ] ],
						"v": [ [ 0.271, -1.58 ], [ -1.021, 1.58 ], [ -0.273, 1.58 ], [ 1.022, -1.58 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.999999940162, 0.999999940162, 1, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 80.139, 50.772 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 11",
			"np": 2,
			"cix": 2,
			"ix": 11,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0 ] ],
						"o": [ [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0 ] ],
						"v": [ [ 0.04, -1.58 ], [ -1.253, 1.58 ], [ -0.043, 1.58 ], [ 1.253, -1.58 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.999999940162, 0.999999940162, 1, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 64.658, 50.772 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 12",
			"np": 2,
			"cix": 2,
			"ix": 12,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, -3.413 ], [ -3.432, 0 ], [ 0, 3.413 ], [ 3.424, 0 ] ],
						"o": [ [ 0, 3.413 ], [ 3.424, 0 ], [ 0, -3.413 ], [ -3.432, 0 ] ],
						"v": [ [ -6.205, 0 ], [ 0.004, 6.183 ], [ 6.205, 0 ], [ 0.004, -6.183 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.999999940162, 0.999999940162, 1, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 72.65, 49.627 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 13",
			"np": 2,
			"cix": 2,
			"ix": 13,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0, -1.153 ], [ 0, 0 ], [ -1.134, 0 ], [ 0, 0 ], [ 0, 1.154 ], [ 0, 0 ], [ 1.135, 0 ] ],
						"o": [ [ -1.134, 0 ], [ 0, 0 ], [ 0, 1.154 ], [ 0, 0 ], [ 1.135, 0 ], [ 0, 0 ], [ 0, -1.153 ], [ 0, 0 ] ],
						"v": [ [ -12.713, -9.036 ], [ -14.766, -6.948 ], [ -14.766, 6.948 ], [ -12.713, 9.036 ], [ 12.712, 9.036 ], [ 14.766, 6.948 ], [ 14.766, -6.948 ], [ 12.712, -9.036 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.864472572476, 0.033036635904, 0.029038397471, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 72.516, 48.792 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 14",
			"np": 2,
			"cix": 2,
			"ix": 14,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ -11.336, -8.518 ], [ 0, 2.588 ], [ 6.044, 0 ] ],
						"o": [ [ 1.711, 2.778 ], [ 11.258, -8.389 ], [ -7.098, -4.129 ], [ -12.762, -0.001 ] ],
						"v": [ [ -20.941, -9.438 ], [ -0.206, 15.688 ], [ 20.94, -10.155 ], [ 0.937, -15.686 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.999999940162, 0.999999940162, 1, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 48.742, 20.938 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 15",
			"np": 2,
			"cix": 2,
			"ix": 15,
			"mn": "ADBE Vector Group",
			"hd": false
		}, {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0, 0 ], [ 0, -15.285 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 7.135, 0 ] ],
						"o": [ [ 0, 0 ], [ -7.137, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, 0 ], [ 0, -15.285 ], [ 0, 0 ] ],
						"v": [ [ -0.208, -36.147 ], [ -13.441, -36.147 ], [ -44.297, -4.07 ], [ -44.297, 36.147 ], [ 44.297, 36.147 ], [ 44.297, -4.07 ], [ 13.029, -36.147 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.078990988638, 0.191951183244, 0.33772017535, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 49.547, 44.684 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 16",
			"np": 2,
			"cix": 2,
			"ix": 16,
			"mn": "ADBE Vector Group",
			"hd": false
		} ],
		"ip": 0,
		"op": 900,
		"st": 0,
		"bm": 0
	}, {
		"ddd": 0,
		"ind": 6,
		"ty": 4,
		"nm": "hår Outlines",
		"sr": 1,
		"ks": {
			"o": { "a": 0, "k": 100, "ix": 11 },
			"r": { "a": 0, "k": 0, "ix": 10 },
			"p": { "a": 0, "k": [ 50, 54.691, 0 ], "ix": 2 },
			"a": { "a": 0, "k": [ 47.758, 53.94, 0 ], "ix": 1 },
			"s": { "a": 0, "k": [ 100, 100, 100 ], "ix": 6 }
		},
		"ao": 0,
		"shapes": [ {
			"ty": "gr",
			"it": [ {
				"ind": 0,
				"ty": "sh",
				"ix": 1,
				"ks": {
					"a": 0,
					"k": {
						"i": [ [ 0, 0 ], [ 0.125, -26.378 ], [ -40.694, -0.868 ], [ -0.298, -0.008 ], [ -0.422, -0.001 ], [ -0.319, 0.004 ], [ -0.303, 0.012 ], [ -0.035, 22.223 ], [ 23.454, 0.038 ] ],
						"o": [ [ -23.415, 0.014 ], [ -0.035, 22.225 ], [ 0.29, 0.011 ], [ 0.334, 0.003 ], [ 0.327, 0 ], [ 0.313, -0.006 ], [ 40.697, -0.739 ], [ 0.041, -26.411 ], [ 0, 0 ] ],
						"v": [ [ 0.111, -48.691 ], [ -42.432, -0.939 ], [ -1.932, 48.65 ], [ -1.033, 48.669 ], [ 0.038, 48.691 ], [ 0.988, 48.673 ], [ 1.928, 48.655 ], [ 42.583, -0.803 ], [ 0.191, -48.691 ] ],
						"c": true
					},
					"ix": 2
				},
				"nm": "Path 1",
				"mn": "ADBE Vector Shape - Group",
				"hd": false
			}, {
				"ty": "fl",
				"c": { "a": 0, "k": [ 0.169087578269, 0.127239840638, 0.109252144309, 1 ], "ix": 4 },
				"o": { "a": 0, "k": 100, "ix": 5 },
				"r": 1,
				"nm": "Fill 1",
				"mn": "ADBE Vector Graphic - Fill",
				"hd": false
			}, {
				"ty": "tr",
				"p": { "a": 0, "k": [ 47.683, 53.94 ], "ix": 2 },
				"a": { "a": 0, "k": [ 0, 0 ], "ix": 1 },
				"s": { "a": 0, "k": [ 100, 100 ], "ix": 3 },
				"r": { "a": 0, "k": 0, "ix": 6 },
				"o": { "a": 0, "k": 100, "ix": 7 },
				"sk": { "a": 0, "k": 0, "ix": 4 },
				"sa": { "a": 0, "k": 0, "ix": 5 },
				"nm": "Transform"
			} ],
			"nm": "Group 1",
			"np": 2,
			"cix": 2,
			"ix": 1,
			"mn": "ADBE Vector Group",
			"hd": false
		} ],
		"ip": 0,
		"op": 900,
		"st": 0,
		"bm": 0
	} ],
	"markers": []
};

class EllaBlunk extends React.Component<EllaProps, any> {

	constructor(props: EllaProps) {
		super(props);
	}

	render() {

		const defaultOptions = {
			loop: true,
			autoplay: true,
			animationData: blunkedamedata,
			rendererSettings: {
				preserveAspectRatio: 'xMidYMid slice'
			}
		};

		const blunkeDameAspectRatio = 100/168;
		const hoyde = this.props.hoyde | 168;
		const bredde = hoyde * blunkeDameAspectRatio;

		return (
			<div className="ellablunk_bakgrunn_wrapper">
				<div className="ellablunk_bakgrunn">
					<Lottie
						options={defaultOptions}
						height={hoyde}
						width={bredde}
						isStopped={this.props.erStoppet}
						isPaused={this.props.erPauset}
					/>
				</div>
			</div>
		)
	}
}

(EllaBlunk as React.ComponentClass).propTypes = {
	/**
	 * Høyde i piksler
	 */
	hoyde: PT.number,
	/**
	 * Animasjon er stoppet
	 */
	erStoppet: PT.bool,
	/**
	 * Animasjon er
	 */
	erPauset: PT.bool,
};

export default EllaBlunk;
