// nav-frontend-grid/src/column.js
import React, { PropTypes as PT } from 'react';
import classNames from 'classnames';
import 'nav-frontend-grid-style';// eslint-disable-line import/extensions

const cls = (className, xs, sm, md, lg) => classNames('col', className, {
    [`col-xs-${xs}`]: !!xs,
    [`col-sm-${sm}`]: !!sm,
    [`col-md-${md}`]: !!md,
    [`col-lg-${lg}`]: !!lg
});


function Column({ children, className, xs, sm, md, lg, ...props }) {
    return (
        <div className={cls(className, xs, sm, md, lg)} {...props}>{children}</div>
    );
}

Column.defaultProps = {
    className: undefined,
    xs: undefined,
    sm: undefined,
    md: undefined,
    lg: undefined
};

Column.propTypes = {
    className: PT.string,
    xs: PT.oneOf(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
    sm: PT.oneOf(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
    md: PT.oneOf(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
    lg: PT.oneOf(['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']),
    children: PT.oneOfType([
        PT.arrayOf(PT.node),
        PT.node
    ]).isRequired
};

export default Column;
