// nav-frontend-lenkepanel/src/index.js
import React, { Component, PropTypes as PT } from 'react';
import classnames from 'classnames';
import Typografi from 'nav-frontend-typografi';
import 'nav-frontend-lenkepanel-style'; // eslint-disable-line import/extensions

const cls = (className) => classnames('lenkepanel', className);
/**
 * Lenkepanel
 */
class Lenkepanel extends Component {
    render() {
        const { className, children, tittelProps, linkCreator, ...renderProps } = this.props;
        const headingConfig = { type: tittelProps, tag: 'span' };
        const heading =
            <Typografi key="heading" {...headingConfig} className="lenkepanel__heading">{children}</Typografi>;

        return linkCreator({
            ...renderProps,
            className: cls(className),
            children: [
                heading,
                <span key="indikator" className="lenkepanel__indikator" />
            ]
        });
    }
}

Lenkepanel.propTypes = {
    /**
     * Klassenavn
     */
    className: PT.string,
    /**
     * Lenke (denne sendes som prop til linkCreator)
     */
    href: PT.string.isRequired,
    /**
     * Tekstinnhold for lenkepanel
     */
    children: PT.oneOfType([
        PT.arrayOf(PT.node),
        PT.node
    ]).isRequired,
    /**
     * Typografitype
     */
    tittelProps: PT.string,
    /**
     * Funksjon som brukes for å lage lenken. her kan du spessifisere din egen hvis du f.eks. bruker React Router
     */
    linkCreator: PT.func
};
Lenkepanel.defaultProps = {
    className: undefined,
    tittelProps: 'undertittel',
    linkCreator: (props) => <a {...props} /> // eslint-disable-line jsx-a11y/anchor-has-content
};

export default Lenkepanel;
