// nav-frontend-lenkepanel/src/lenkepanel.js
import PT from 'prop-types';
import React, { Component } from 'react';
import Typografi from 'nav-frontend-typografi';
import 'nav-frontend-lenkepanel-style'; // eslint-disable-line import/extensions
import LenkepanelInnhold from './lenkepanelinnhold'; // eslint-disable-line import/extensions

/**
 * Lenkepanel
 */
class Lenkepanel extends Component { // eslint-disable-line react/prefer-stateless-function
    render() {
        const { children, tittelProps, ...props } = this.props;
        const headingConfig = { type: tittelProps, tag: 'span' };
        const heading = (
            <Typografi
                key="heading"
                {...headingConfig}
                className="lenkepanel__heading"
            >
                { children }
            </Typografi>
        );
        return (
            <LenkepanelInnhold {...props}>
                { heading }
                { <span key="indikator" className="lenkepanel__indikator" /> }
            </LenkepanelInnhold>
        );
    }
}

Lenkepanel.propTypes = {
    /**
     * Klassenavn
     */
    className: PT.string,
    /**
     * Typografiklasse
     */
    tittelProps: PT.string.isRequired,
    /**
     * Lenke (denne sendes som prop til linkCreator)
     */
    href: PT.string.isRequired,
    /**
     * Innhold for lenkepanel
     */
    children: PT.string.isRequired,
    /**
     * Funksjon som brukes for å lage lenken. her kan du spesifisere din egen hvis du f.eks. bruker React Router
     */
    linkCreator: PT.func
};

Lenkepanel.defaultProps = {
    className: undefined,
    linkCreator: (props) => <a {...props} /> // eslint-disable-line jsx-a11y/anchor-has-content
};

export default Lenkepanel;
