// nav-frontend-lenkepanel/src/Lenkepanel-base.js
import PT from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import 'nav-frontend-lenkepanel-style'; // eslint-disable-line import/extensions

const cls = (className) => classnames('lenkepanel', className);
/**
 * Lenkepanel
 */
class LenkepanelBase extends Component {
    render() {
        const { className, children, linkCreator, ...renderProps } = this.props;

        return linkCreator({
            ...renderProps,
            className: cls(className),
            children: [
                children,
                <span key="indikator" className="lenkepanel__indikator" />
            ]
        });
    }
}

LenkepanelBase.propTypes = {
    /**
     * Klassenavn
     */
    className: PT.string,
    /**
     * Lenke (denne sendes som prop til linkCreator)
     */
    href: PT.string.isRequired,
    /**
     * Tekstinnhold for lenkepanel
     */
    children: PT.oneOfType([
        PT.arrayOf(PT.node),
        PT.node
    ]).isRequired,
    /**
     * Funksjon som brukes for å lage lenken. her kan du spessifisere din egen hvis du f.eks. bruker React Router
     */
    linkCreator: PT.func
};
LenkepanelBase.defaultProps = {
    className: undefined,
    linkCreator: (props) => <a {...props} /> // eslint-disable-line jsx-a11y/anchor-has-content
};

export default LenkepanelBase;
