// nav-frontend-lenkepanel/src/Lenkepanel-base.tsx
import * as PT from 'prop-types';
import * as React from 'react';
import * as classNames from 'classnames';
import 'nav-frontend-lenkepanel-style';

const cls = (className) => classNames('lenkepanel', className);

export interface LenkepanelBaseProps {
    className?: string;
    href: string;
    children: React.ReactNode | React.ReactNode[];
    linkCreator?: Function;
}



class LenkepanelBase extends React.Component<LenkepanelBaseProps>{
    render() {
        const { className, children, linkCreator = (props) => <a {...props}/>, ...renderProps } = this.props;

        return linkCreator({
            ...renderProps,
            className: cls(className),
            children: [
                children,
                <span key="indikator" className="lenkepanel__indikator" />
            ]
        });
    }
}


(LenkepanelBase as React.ComponentClass).propTypes = {
    /**
     * Klassenavn
     */
    className: PT.string,
    /**
     * Lenke (denne sendes som prop til linkCreator)
     */
    href: PT.string.isRequired,
    /**
     * Tekstinnhold for lenkepanel
     */
    children: PT.oneOfType([
        PT.arrayOf(PT.node),
        PT.node
    ]).isRequired,
    /**
     * Funksjon som brukes for å lage lenken. her kan du spessifisere din egen hvis du f.eks. bruker React Router
     */
    linkCreator: PT.func
};

(LenkepanelBase as React.ComponentClass).defaultProps = {
    className: undefined,
    linkCreator: (props) => <a {...props} />
};

export default LenkepanelBase;
