// nav-frontend-veileder-panel/src/veileder-panel.js
import React from 'react';
import PT from 'prop-types';
import cn from 'classnames';

import Panel from 'nav-frontend-paneler';
import Veileder from 'nav-frontend-veileder';
import "nav-frontend-veileder-panel-style";

const cls = (type, compact) => cn('veileder-panel', {
    'veileder-panel--plakat': type === 'plakat',
    'veileder-panel--compact': compact
});

class VeilederPanel extends React.Component {
    render() {
        const { children, svg, center, color, type, compact, ...other } = this.props;
        const width = (type === 'plakat') ? 100 : 80 ;
        return (
            <div className={cls(type, compact)}>
                <Veileder width={width} center={center} color={color}>
                    {svg}
                </Veileder>
                <Panel style={{borderColor: color}}>
                    {children}
                </Panel>
            </div>
        );
    }
}

VeilederPanel.propTypes = {
    children: PT.node.isRequired,
    svg: PT.node.isRequired,
    center: PT.bool,
    color: PT.string,
    type: PT.oneOf(['normal', 'plakat'])
};

VeilederPanel.defaultProps = {
    center: false,
    color: '#E9E7E7',
    type: 'normal'
};

export default VeilederPanel;
