// nav-frontend-veileder-panel/src/veileder-panel.tsx
import * as React from 'react';
import * as PT from 'prop-types';
import * as cn from 'classnames';

import Panel from 'nav-frontend-paneler';
import Veileder from 'nav-frontend-veileder';
import "nav-frontend-veileder-panel-style";

const cls = (type, compact) => cn('veileder-panel', {
    'veileder-panel--plakat': type === 'plakat',
    'veileder-panel--compact': compact
});

export interface VeilederPanelProps {
    children: React.ReactNode | React.ReactChild | React.ReactChildren;
    svg: React.ReactNode | React.ReactChild | React.ReactChildren;
    center?: boolean;
    color?: string;
    type: 'normal' | 'plakat';
    compact?: boolean;
};

class VeilederPanel extends React.Component<VeilederPanelProps> {
    render() {
        const { children, svg, center, color, type, compact, ...other } = this.props;
        const width = (type === 'plakat') ? 100 : 80 ;
        return (
            <div className={cls(type, compact)} {...other}>
                <Veileder width={width} center={center} color={color}>
                    {svg}
                </Veileder>
                <Panel style={{borderColor: color}}>
                    {children}
                </Panel>
            </div>
        );
    }
}

(VeilederPanel as any).propTypes = {
    /**
     * Tekst eller markup som definerer innholdet i panelet
     */
    children: PT.node.isRequired,
    /**
     * Custom SVG som videreformidles som child til Veileder-komponenten
     */
    svg: PT.node.isRequired,
    /**
     * Videreformidles til Veileder-komponenten (se Veileder docs)
     */
    center: PT.bool,
    /**
     * Bestemmer farge på panel-border og bakgrunnsfarge på Veileder-komponenten
     */
    color: PT.string,
    /**
     * Bestemmer om panelet skal rendres som liten (normal) eller stor (plakat)
     */
    type: PT.oneOf(['normal', 'plakat']),
    /**
     * Bestemmer om Veileder-komponenten skal rendres inne i panelet (false, default), eller oppå border (true)
     */
    compact: PT.bool
};

(VeilederPanel as any).defaultProps = {
    center: false,
    color: '#E9E7E7',
    type: 'normal',
    compact: false
};

export default VeilederPanel;
