// nav-frontend-inngang/src/dobbelingressinngang.js
import React, { Component } from 'react';
import PT from 'prop-types';
import { Systemtittel, Ingress } from 'nav-frontend-typografi';
import { HoyreChevron } from 'nav-frontend-chevron';
import Ikon from 'nav-frontend-ikoner-assets'; // eslint-disable-line import/extensions
import 'nav-frontend-inngang-style'; // eslint-disable-line import/extensions

class DobbelIngressInngang extends Component { // eslint-disable-line react/prefer-stateless-function
    render() {
        const { ingress1, ingress2, tittel, ikon, ...props } = this.props;

        return (
            <a {... props} className="inngang inngang__dobbelingress">
                <div className="inngang__dobbelingress__content">
                    <div className="ikonWrapper">
                        <Ikon kind={ikon} size={100} />
                    </div>
                    <div className="textWrapper">
                        <Systemtittel>{ tittel }</Systemtittel>
                        <Ingress>
                            { ingress1 }
                            <HoyreChevron />
                        </Ingress>
                        <Ingress>
                            { ingress2 }
                            <HoyreChevron />
                        </Ingress>
                    </div>
                </div>
            </a>
        );
    }
}

DobbelIngressInngang.propTypes = {
    tittel: PT.string.isRequired,
    ingress1: PT.string.isRequired,
    ingress2: PT.string.isRequired,
    ikon: PT.string.isRequired
};

export default DobbelIngressInngang;
