// nav-frontend-inngang/src/normalvertikalinngang.js
import React, { Component } from 'react';
import PT from 'prop-types';
import { Undertittel, Normaltekst } from 'nav-frontend-typografi';
import { HoyreChevron } from 'nav-frontend-chevron';
import Ikon from 'nav-frontend-ikoner-assets'; // eslint-disable-line import/extensions
import 'nav-frontend-inngang-style'; // eslint-disable-line import/extensions
import 'nav-frontend-lenker-style'; // eslint-disable-line import/extensions

class NormalVertikalInngang extends Component { // eslint-disable-line react/prefer-stateless-function

    static renderHoyreChevron() {
        return (
            <div className="ikonWrapper">
                <HoyreChevron />
            </div>
        );
    }

    renderTekstInnhold(tittel, tekst) { // eslint-disable-line class-methods-use-this
        return (
            <div className="tekstWrapper">
                <Undertittel>{ tittel }</Undertittel>
                <Normaltekst>{ tekst }</Normaltekst>
            </div>
        );
    }

    renderIkon(ikon) { // eslint-disable-line class-methods-use-this
        return (
            <div className="ikonWrapper">
                <Ikon kind={ikon} />
            </div>
        );
    }

    renderNormalVertikalInngangMedIkon(ikon, tittel, tekst) {
        return (
            <div className="inngang__normalvertikalinngang--medikon">
                { this.renderIkon(ikon) }
                { this.renderTekstInnhold(tittel, tekst) }
                { NormalVertikalInngang.renderHoyreChevron() }
            </div>
        );
    }

    renderNormalVertikalInngangMedLink(tittel, tekst, link) {
        return (
            <div className="inngang__normalvertikalinngang--medlink">
                { this.renderTekstInnhold(tittel, tekst) }
                <a className="lenke" href={link}>Les mer</a>
            </div>
        );
    }

    renderNormalVertikalInngangDefault(tittel, tekst) {
        return (
            <div>
                { NormalVertikalInngang.renderHoyreChevron() }
                { this.renderTekstInnhold(tittel, tekst) }
            </div>
        );
    }

    render() {
        const { tekst, tittel, ikon, lesMerHref, ...props } = this.props;
        let renderedContent = null;

        if (ikon !== false) {
            renderedContent = this.renderNormalVertikalInngangMedIkon(ikon, tittel, tekst);
        } else if (lesMerHref !== false) {
            renderedContent = this.renderNormalVertikalInngangMedLink(tittel, tekst, lesMerHref);
        } else {
            renderedContent = this.renderNormalVertikalInngangDefault(tittel, tekst);
        }


        return (
            <div {... props} className="inngang inngang__normalvertikalinngang" tabIndex={0}>
                { renderedContent }
            </div>
        );
    }
}

NormalVertikalInngang.defaultProps = {
    ikon: false,
    lesMerHref: false
};

NormalVertikalInngang.propTypes = {
    tittel: PT.string.isRequired,
    tekst: PT.string.isRequired,
    ikon: PT.string,
    lesMerHref: PT.string
};

export default NormalVertikalInngang;
