// nav-frontend-knapper/src/index.js
import React, { Component, PropTypes as PT } from 'react';
import classNames from 'classnames';
import { omit } from 'nav-frontend-js-utils';
import 'nav-frontend-knapper-style'; // eslint-disable-line import/extensions


export const cls = (className, props) => classNames('knapp', className, {
    'knapp--hoved': props.type === 'hoved',
    'knapp--fare': props.type === 'fare',
    'knapp--standard': props.type === 'standard',
    'knapp--mini': props.mini,
    // eslint-disable-next-line no-use-before-define
    'knapp--disabled': props.disabled,
    'knapp--spinner': props.spinner
});

function skalVareDisabled(props) {
    if (props.autoDisableVedSpinner) { // eslint-disable-line no-use-before-define
        return props.disabled || props.spinner;
    }
    return props.disabled;
}
/**
 *  Knapp, Hovedknapp og Fareknapp arver alle fra KnappBase, og har da samme props.
 *  Unntaket er `type` som blir satt av de respektive underklassene.
 */
class KnappBase extends Component { // eslint-disable-line react/prefer-stateless-function
    render() {
        const { children, className, ...props } = this.props;

        const spinner = props.spinner ? <span className="knapp__spinner" /> : null;
        const disabled = skalVareDisabled(props);
        const domProps = omit(
            { ...props, disabled },
            'type',
            'mini',
            'spinner',
            'autoDisableVedSpinner',
            'inaktivKlasseVedDisabled'
        );

        return (
            <button role="button" className={cls(className, props)} {...domProps}>
                {children}
                {spinner}
            </button>
        );
    }
}

KnappBase.defaultProps = {
    autoDisableVedSpinner: false,
    inaktivKlasseVedDisabled: false,
    className: null,
    mini: false,
    disabled: false,
    spinner: false
};

KnappBase.propTypes = {
    children: PT.node.isRequired,
    className: PT.string,
    type: PT.oneOf(['standard', 'hoved', 'fare']).isRequired,
    mini: PT.bool,
    disabled: PT.bool,
    spinner: PT.bool,
    autoDisableVedSpinner: PT.bool,
    inaktivKlasseVedDisabled: PT.bool
};

export default KnappBase;

export { default as Knapp } from './knapp';
export { default as Hovedknapp } from './hovedknapp';
export { default as Fareknapp } from './fareknapp';
