// nav-frontend-telefoninput/src/telefoninput-types.ts
import { NavFrontendInputProps } from 'nav-frontend-skjema/src/input';

export interface Country {
    name: string;
    regions: string[];
    iso2: string;
    dialCode: string;
    format?: string;
    priority?: string;
    hasAreaCodes?: boolean;
}

export interface TelefonInputProps {
    /**
     * -
     */
    label: string;

    /**
     * -
     */
    keys?: {[key: string]:  number };
    /**
     * -
     */
    excludeCountries: string[];
    /**
     * -
     */
    onlyCountries: string[];
    /**
     * -
     */
    preferredCountries: string[];
    /**
     * -
     */
    defaultCountry: string;

    /**
     * -
     */
    value: string;
    /**
     * -
     */
    placeholder: string;
    /**
     * -
     */
    searchPlaceholder?: string;
    /**
     * -
     */
    disabled?: boolean;

    /**
     * -
     */
    autoFormat: boolean;
    /**
     * -
     */
    disableAreaCodes: boolean;
    /**
     * -
     */
    disableCountryCode: boolean;
    /**
     * -
     */
    disableDropdown: boolean;
    /**
     * -
     */
    enableLongNumbers: boolean;
    /**
     * -
     */
    countryCodeEditable: boolean;
    /**
     * -
     */
    enableSearchField: boolean;

    /**
     * -
     */
    regions: string | string[];

    /**
     * -
     */
    inputExtraProps: NavFrontendInputProps;
    /**
     * -
     */
    localization: any;
    /**
     * -
     */
    masks: any;

    /**
     * -
     */
    onEnterKeyPress: any;
    /**
     * -
     */
    onChange: any;
    /**
     * -
     */
    onFocus: any;
    /**
     * -
     */
    onBlur: any;
    /**
     * -
     */
    onClick: any;
    /**
     * -
     */
    onKeyDown: any;
}
