// nav-frontend-lesmerpanel/src/lesmerpanel.tsx
import * as React from 'react';
import { PropTypes as PT } from 'prop-types';
import * as classNames from 'classnames';
import 'nav-frontend-lesmerpanel-style';
import LesmerpanelToggle from './lesmerpanelToggle';
import { scrollTo, erSynligIViewport } from './lesmerutils';

export interface LesMerPanelBaseProps {
    /**
     * Hvorvidt panelet initielt er åpent
     */
    erApen?: boolean;
    /**
     * Funksjon som kalles når panelet åpnes
     */
    onOpen?: (event: React.SyntheticEvent<HTMLButtonElement>) => void;
    /**
     * Funksjon som kalles når panelet lukkes
     */
    onClose?: (event: React.SyntheticEvent<HTMLButtonElement>) => void;
    /**
     * Innhold som plasseres før "Åpne"
     */
    intro: React.ReactNode;
    /**
     * Innhold som vises når man klikker på "Åpne"
     */
    children: React.ReactNode;
    /**
     * ID til panelet
     */
    id?: string;
    /**
     * Tekst som vises for å lukke panelet
     */
    lukkTekst: React.ReactNode;
    /**
     * Tekst som vises for å åpne panelet
     */
    apneTekst: React.ReactNode;
    /**
     * Egendefinert klassenavn
     */
    className?: string;
}

export interface LesMerPanelBaseState {
    erApen: boolean;
    hindreToggle: boolean;
    hoyde: any;
    visInnhold: boolean;
    harTransisjon: boolean;
    animerer: boolean;
}

/**
 * Lesmerpanel
 */
class Lesmerpanel extends React.Component<LesMerPanelBaseProps, LesMerPanelBaseState> {
    lesMerPanel: HTMLDivElement | null;

    mer: HTMLDivElement | null;

    static defaultProps: Partial<LesMerPanelBaseProps> = {
        erApen: false,
        apneTekst: 'Åpne',
        lukkTekst: 'Lukk',
        onClose: () => {},
        onOpen: () => {}
    };

    constructor(props) {
        super(props);
        this.state = {
            erApen: props.erApen,
            animerer: false,
            hindreToggle: false,
            hoyde: !props.erApen ? '0' : 'auto',
            visInnhold: props.erApen,
            harTransisjon: false
        };
        this.onTransitionEnd = this.onTransitionEnd.bind(this);
    }

    onTransitionEnd() {
        if (this.state.harTransisjon) {
            this.setState({
                harTransisjon: false
            });
            if (this.state.erApen) {
                scrollTo(this.lesMerPanel, 600);
                this.setState({
                    hindreToggle: false
                });
                this.setAutoHoyde();
                if (this.mer) {
                    this.mer.focus();
                }
            } else {
                this.setState({
                    hindreToggle: false,
                    visInnhold: false
                });
                if (!erSynligIViewport(this.lesMerPanel)) {
                    scrollTo(this.lesMerPanel, 600);
                }
            }
        }
    }

    setAutoHoyde() {
        /* Fjerner animasjonsklassen slik at Safari ikke
        tegner komponenten på nytt når høyde settes til 'auto': */
        this.setState(
            {
                animerer: false
            },
            () => {
                this.setState({
                    hoyde: 'auto',
                    animerer: false
                });
            });
    }

    apne() {
        this.setState({
            hoyde: '0',
            hindreToggle: true,
            animerer: true,
            visInnhold: true,
            harTransisjon: true
        });
        setTimeout(
            () => {
                const hoyde = this.mer ? this.mer.offsetHeight : '0';
                this.setState({
                    hoyde,
                    erApen: true
                });
            },
            10);
    }

    lukk() {
        const hoyde = this.mer ? this.mer.offsetHeight : '0';
        this.setState({
            hoyde,
            hindreToggle: true,
            harTransisjon: true
        });
        setTimeout(
            () => {
                this.setState({
                    animerer: true,
                    hoyde: '0',
                    erApen: false
                });
            },
            10);
    }

    toggle(e, onOpen, onClose) {
        e.preventDefault();
        if (!this.state.hindreToggle) {
            /* hindreToggle for å hindre dobbelklikk,
            eller at noen klikker mens animasjonen pågår.
            Dobbelklikk vil skape kluss med logikken. */
            if (this.state.erApen) {
                this.lukk();
                onClose();
            } else {
                this.apne();
                onOpen();
            }
        }
    }

    render() {
        const { className, intro, id, children, onOpen, onClose } = this.props;
        const merContainerClassName = classNames('lesMerPanel__merContainer', {
            'lesMerPanel__merContainer--medAnimasjon': this.state.animerer,
        });
        return (
            <div
                id={id}
                ref={(c) => {
                    this.lesMerPanel = c;
                }}
                className={`lesMerPanel ${className}`}
            >
                <div>
                    {intro}
                </div>
                <div
                    style={{ height: this.state.hoyde }}
                    className={merContainerClassName}
                    onTransitionEnd={this.onTransitionEnd}
                >
                    <div
                        className="lesMerPanel__mer"
                        tabIndex={-1}
                        ref={(c) => {
                            this.mer = c;
                        }}
                    >
                        {
                            this.state.visInnhold
                                ? children
                                : null
                        }
                    </div>
                </div>
                <LesmerpanelToggle
                    apneTekst={this.props.apneTekst}
                    lukkTekst={this.props.lukkTekst}
                    erApen={this.state.erApen}
                    onClick={(event) => {
                        this.toggle(event, onOpen, onClose);
                    }}
                />
            </div>);
    }
}

(Lesmerpanel as any).propTypes = {
    /**
     * Hvorvidt panelet initielt er åpent
     */
    erApen: PT.bool,
    /**
     * Funksjon som kalles når panelet åpnes
     */
    onOpen: PT.func,
    /**
     * Funksjon som kalles når panelet lukkes
     */
    onClose: PT.func,
    /**
     * Innhold som plasseres før "Åpne"
     */
    intro: PT.node,
    /**
     * Innhold som vises når man klikker på "Åpne"
     */
    children: PT.node,
    /**
     * ID til panelet
     */
    id: PT.string,
    /**
     * Tekst som vises for å lukke panelet
     */
    lukkTekst: PT.string,
    /**
     * Tekst som vises for å åpne panelet
     */
    apneTekst: PT.string,
    /**
     * Egendefinert klassenavn
     */
    className: PT.string,
};

export default Lesmerpanel;
