// nav-frontend-ekspanderbartpanel/src/ekspanderbartpanel-pure.js
import React, { Component } from 'react';
import PT from 'prop-types';
import Typografi from 'nav-frontend-typografi';
import EkspanderbartpanelBasePure from './ekspanderbartpanel-base-pure';

class EkspanderbartpanelPure extends Component {

    render() {
        const { tittel, tittelProps, ...renderProps } = this.props;
        const headingConfig = { type: tittelProps, tag: 'span' };

        const heading = <Typografi {...headingConfig} className="ekspanderbartPanel__heading">{tittel}</Typografi>;

        return (
            <EkspanderbartpanelBasePure {...renderProps} heading={heading} ariaTittel={tittel} />
        );
    }
}

EkspanderbartpanelPure.propTypes = {
    className: PT.string,
    onClick: PT.func.isRequired,
    tittel: PT.string.isRequired,
    tittelProps: PT.string,
    apen: PT.bool.isRequired,
    children: PT.oneOfType([
        PT.arrayOf(PT.node),
        PT.node
    ]).isRequired
};
EkspanderbartpanelPure.defaultProps = {
    tittelProps: 'undertittel',
    className: undefined
};

export default EkspanderbartpanelPure;
