// nav-frontend-ekspanderbartpanel/src/ekspanderbartpanel-pure.js
import React, { PropTypes as PT } from 'react';
import classnames from 'classnames';
import Collapse from 'react-collapse';
import Typografi from 'nav-frontend-typografi';
import 'nav-frontend-ekspanderbartpanel-style'; // eslint-disable-line import/extensions

const cls = (className, props) => classnames('ekspanderbartPanel', className, {
    'ekspanderbartPanel--lukket': !props.apen,
    'ekspanderbartPanel--apen': props.apen
});

function EkspanderbartpanelPure(props) {
    const { className, children, apen, tittel, tittelProps, onClick, ...renderProps } = props;
    const headingConfig = typeof tittelProps === 'string' ? { type: tittelProps } : tittelProps;

    const heading = <Typografi {...headingConfig} className="ekspanderbartPanel__heading">{tittel}</Typografi>;

    return (
        <div className={cls(className, props)} {...renderProps}>
            <button className="ekspanderbartPanel__hode" onClick={onClick} aria-expanded={apen}>
                {heading}
                <span className="ekspanderbartPanel__indikator" />
            </button>
            <Collapse isOpened={apen}>
                <div className="ekspanderbartPanel__innhold">{children}</div>
            </Collapse>
        </div>
    );
}

EkspanderbartpanelPure.propTypes = {
    className: PT.string,
    onClick: PT.func.isRequired,
    tittel: PT.string.isRequired,
    tittelProps: PT.oneOfType([PT.string, PT.shape({
        type: PT.string,
        tag: PT.string
    })]),
    apen: PT.bool.isRequired,
    children: PT.oneOfType([
        PT.arrayOf(PT.node),
        PT.node
    ]).isRequired
};
EkspanderbartpanelPure.defaultProps = {
    tittelProps: 'undertittel',
    className: undefined
};

export default EkspanderbartpanelPure;
