// nav-frontend-ekspanderbartpanel/src/ekspanderbartpanel-base-pure.tsx
import * as React from 'react';
import * as classnames from 'classnames';
import { UnmountClosed, Collapse, CollapseProps } from 'react-collapse';
import 'nav-frontend-ekspanderbartpanel-style';
import { guid } from 'nav-frontend-js-utils';

export interface EkspanderbartpanelBasePureProps {
    heading: React.ReactNode;
    className?: string;
    onClick: (event: React.SyntheticEvent<HTMLButtonElement>) => void;
    ariaTittel: string;
    apen: boolean;
    children?: React.ReactNode;
    collapseProps?: Partial<CollapseProps>;
    border?: boolean;
    renderContentWhenClosed?: boolean;
}

const cls = (apen: boolean, border: boolean, className?: string) =>
    classnames('ekspanderbartPanel', className, {
        'ekspanderbartPanel--lukket': !apen,
        'ekspanderbartPanel--apen': apen,
        'ekspanderbartPanel--border': border
    });

class EkspanderbartpanelBasePure extends React.Component<EkspanderbartpanelBasePureProps, {}> {
    componentId: string;
    static defaultProps: Partial<EkspanderbartpanelBasePureProps> = {
        border: false
    };
    private isCloseAnimation: boolean = false;

    constructor(props: EkspanderbartpanelBasePureProps) {
        super(props);
        this.componentId = guid();
    }

    componentWillReceiveProps(nextProps) {
        if (this.props.apen && !nextProps.apen) {
            this.isCloseAnimation = true;
        }
    }

    onRestProxy = () => {
        this.isCloseAnimation = false;

        const { collapseProps } = this.props;
        if (collapseProps && collapseProps.onRest) {
            collapseProps.onRest();
        }
    }

    tabHandler(event) {
        const { keyCode } = event;
        const isTab = keyCode === 9;

        if (isTab && this.isCloseAnimation) {
            event.preventDefault();
        }
    }

    render() {
        const {
            className,
            children,
            apen,
            heading,
            ariaTittel,
            onClick,
            collapseProps,
            border,
            renderContentWhenClosed,
            ...renderProps
        }: EkspanderbartpanelBasePureProps = this.props;

        const hiddenContentId = (collapseProps && collapseProps.id) || this.componentId;
        const CollapseComponent = renderContentWhenClosed ? Collapse : UnmountClosed;
        return (
            <div className={cls(apen, border!, className)} {...renderProps}>
                <button
                    className="ekspanderbartPanel__hode"
                    onKeyDown={(event) => this.tabHandler(event)}
                    onClick={onClick}
                    aria-expanded={apen}
                    aria-controls={renderContentWhenClosed ? hiddenContentId : undefined}
                    type="button"
                >
                    <div className="ekspanderbartPanel__flex-wrapper">
                        {heading}
                        <span className="ekspanderbartPanel__indikator" />
                    </div>
                </button>
                <CollapseComponent
                    id={renderContentWhenClosed ? hiddenContentId : undefined}
                    isOpened={apen}
                    onRest={this.onRestProxy}
                    {...collapseProps}
                >
                    <article aria-label={ariaTittel} className="ekspanderbartPanel__innhold">
                        {children}
                    </article>
                </CollapseComponent>
            </div>
        );
    }
}

export default EkspanderbartpanelBasePure;
