// nav-frontend-ekspanderbartpanel/src/ekspanderbartpanel-base.tsx
import * as React from 'react';
import 'nav-frontend-ekspanderbartpanel-style';
import EkspanderbartpanelBasePure, { EkspanderbartpanelBasePureProps } from './ekspanderbartpanel-base-pure';

export interface EkspanderbartpanelBaseState {
    apen: boolean;
}

class EkspanderbartpanelBase extends React.Component<EkspanderbartpanelBasePureProps, EkspanderbartpanelBaseState> {
    static defaultProps: Partial<EkspanderbartpanelBasePureProps> = {
        apen: false,
        border: false
    };

    constructor(props) {
        super(props);

        this.state = {
            apen: this.props.apen!
        };

        this.handleClick = this.handleClick.bind(this);
    }

    handleClick(event) {
        event.preventDefault();
        this.setState({ apen: !this.state.apen });
        if (this.props.onClick) {
            this.props.onClick(event);
        }
    }

    render() {
        const { ...renderProps } = this.props;
        return (
            <EkspanderbartpanelBasePure
                {...renderProps}
                apen={this.state.apen}
                onClick={this.handleClick}
            />
        );
    }
}

export default EkspanderbartpanelBase;
