// nav-frontend-ekspanderbartpanel/src/ekspanderbartpanel-base.tsx
import * as React from 'react';
import 'nav-frontend-ekspanderbartpanel-style'; // eslint-disable-line import/extensions
import EkspanderbartpanelBasePure from './ekspanderbartpanel-base-pure';

export interface EkspanderbartpanelBaseProps {
    apen?: boolean;
    onClick: (event: React.SyntheticEvent<HTMLButtonElement>) => void;
    heading: React.ReactNode;
    ariaTittel: string;
}

export interface EkspanderbartpanelBaseState {
    apen: boolean;
}

class EkspanderbartpanelBase extends React.Component<EkspanderbartpanelBaseProps, EkspanderbartpanelBaseState> {
    constructor(props) {
        super(props);

        this.state = {
            apen: this.props.apen ? this.props.apen : false
        };

        this.handleClick = this.handleClick.bind(this);
    }

    handleClick(event) {
        event.preventDefault();
        this.setState({ apen: !this.state.apen });
        this.props.onClick(event);
    }

    render() {
        const { ...renderProps } = this.props;
        return (
            <EkspanderbartpanelBasePure
                {...renderProps}
                apen={this.state.apen}
                onClick={this.handleClick}
            />
        );
    }
}

export default EkspanderbartpanelBase;
