// nav-frontend-ekspanderbartpanel/src/index.js
import PT from 'prop-types';
import React, { Component } from 'react';
import 'nav-frontend-ekspanderbartpanel-style'; // eslint-disable-line import/extensions
import EkspanderbartpanelPure from './ekspanderbartpanel-pure';

/**
 * Self-contained `Ekspanderbartpanel`.
 * Denne komponenten holder selv styr på om innholdet skal vises eller ikke.
 */
class Ekspanderbartpanel extends Component {
    constructor(props) {
        super(props);

        this.state = {
            apen: this.props.apen
        };

        this.handleClick = this.handleClick.bind(this);
    }

    handleClick(event) {
        event.preventDefault();
        this.setState({ apen: !this.state.apen });
        this.props.onClick(event);
    }

    render() {
        const { apen: _apen, ...renderProps } = this.props;
        return (
            <EkspanderbartpanelPure {...renderProps} apen={this.state.apen} onClick={this.handleClick} />
        );
    }
}

Ekspanderbartpanel.propTypes = {
    /**
     * Skal komponenten være `default` åpen
     */
    apen: PT.bool,
    /**
     * Callback funksjon for når knappen blir klikket på
     */
    onClick: PT.func
};
Ekspanderbartpanel.defaultProps = {
    apen: false,
    onClick: () => {}
};

export default Ekspanderbartpanel;

export { default as EkspanderbartpanelPure } from './ekspanderbartpanel-pure';
