// nav-frontend-ikonknapp/src/ikonknapp.tsx
import * as React from 'react';
import * as PT from 'prop-types';
import * as classnames from 'classnames';

import KnappBase, { KnappBaseProps } from 'nav-frontend-knapper';
import Ikon from 'nav-frontend-ikoner-assets';

import 'nav-frontend-ikonknapp-style';

const ikonKnappCls = (props) => classnames('ikonknapp', {
    'ikonknapp--ikonHøyre': props.ikonHøyre,
    'ikonknapp--kompakt': props.kompakt,
    'ikonknapp--bareIkon': props.bareIkon,
    'ikonknapp--rund': props.rund
});

export interface IkonknappProps extends KnappBaseProps {
    /**
     * Egendefinert klassenavn
     */
    className?: string;
    /**
     * Ikon SVG
     */
    ikon: React.ReactNode;
    /**
     * Bestemmer om ikonet plasseres på høyre side, default er venstre
     */
    ikonHøyre?: false;
    /**
     * Kompakt versjon med lik avstand til ikon fra sidene som over/under
     */
    kompakt?: false;
    /**
     * Variant hvor bare ikonet rendres uten tekst
     */
    bareIkon?: false;
    /**
     * Samme som bareIkon-variant men med avrundede hjørner
     */
    rund?: false;
    
}

class Ikonknapp extends React.Component<IkonknappProps> {
    static defaultProps: Partial<IkonknappProps> = {
        type: 'standard'
    };

    render() {
        const { children, className, ikon, ...other } = this.props;
        return (
            <KnappBase className={ikonKnappCls(this.props)} {...other}>
                <span className="ikonknapp__ikon">{ikon}</span>
                <span className="ikonknapp__tekst">{children}</span>
            </KnappBase>
        );
    }
}

(Ikonknapp as React.ComponentClass).propTypes = {
    /**
     * Egendefinert klassenavn.
     */
    className: PT.string,
    /**
     * Ikon SVG
     */
    ikon: PT.node.isRequired,
    /**
     * Bestemmer om ikonet plasseres på høyre side, default er venstre
     */
    ikonHøyre: PT.bool,
    /**
     * Kompakt versjon med lik avstand til ikon fra sidene som over/under
     */
    kompakt: PT.bool,
    /**
     * Variant hvor bare ikonet rendres uten tekst
     */
    bareIkon: PT.bool,
    /**
     * Samme som bareIkon-variant men med avrundede hjørner
     */
    rund: PT.bool
};

(Ikonknapp as React.ComponentClass).defaultProps = {
    type: 'standard'
};

export default Ikonknapp;
