// nav-frontend-hjelpetekst/src/index.js
import React, { Component, PropTypes as PT } from 'react';
import classNames from 'classnames';
import 'nav-frontend-hjelpetekst-style';// eslint-disable-line import/extensions

const cls = (type, className) => classNames('hjelpetekst', className, {
    'hjelpetekst__under': type === 'under',
    'hjelpetekst__over': type === 'over',
    'hjelpetekst__hojre': type === 'hojre',
    'hjelpetekst__venstre': type === 'venstre'
});

class HjelpetekstBase extends Component {
    constructor(props) {
        super(props);
        this.toggleSynlighet = this.toggleSynlighet.bind(this);
        this.state = {
            erSynlig: false
        };
    }

    toggleSynlighet() {
        event.preventDefault();
        this.setState({
            erSynlig: !this.state.erSynlig
        });
    }

    renderContent(props) {
        return (
            <span className={cls(props.type, classNames)}>{props.children}<a className='hjelpetekst__stangknapp' onClick={this.toggleSynlighet}></a>
            </span>
        );
    }
    render(props) {
        const maybeContent = this.state.erSynlig ? this.renderContent(this.props) : null;
        return (
            <div className="hjelpetekst__container">
                <a className="hjelpetekst__opneknapp" onClick={this.toggleSynlighet} title={this.props.tittel}><span aria-hidden="true">?</span></a>
                {maybeContent}
            </div>
        );
    }
}

HjelpetekstBase.propTypes = {
  children: PT.node.isRequired,
  type: PT.oneOf(['over','under','hojre','venstre']).isRequired,
  className: PT.string
};

HjelpetekstBase.defaultProps = {
    tittel: 'Hjelptekst'
};

export default HjelpetekstBase;

export const HjelpetekstOver = (props) => <HjelpetekstBase type="over" {...props} />;
export const HjelpetekstUnder = (props) => <HjelpetekstBase type="under" {...props} />;
export const HjelpetekstHojre = (props) => <HjelpetekstBase type="hojre" {...props} />;
export const HjelpetekstVenstre = (props) => <HjelpetekstBase type="venstre" {...props} />;
