// nav-frontend-hjelpetekst/src/index.js
import React, { PropTypes as PT } from 'react';
import classNames from 'classnames';
import 'nav-frontend-hjelpetekst-style';// eslint-disable-line import/extensions

const cls = (type, className) => classNames('hjelpetekst', className, {
    'hjelpetekst__under': type === 'under',
    'hjelpetekst__over': type === 'over',
    'hjelpetekst__hojre': type === 'hojre',
    'hjelpetekst__venstre': type === 'venstre'
});

function HjelpetekstBase({ type, className, children, ...props }) {
    return <div className={cls(type, className)} {...props}>{children}</div>;
}

HjelpetekstBase.propTypes = {
    children: PT.node.isRequired,
    type: PT.oneOf(['over','under','hojre','venstre']).isRequired,
    className: PT.string
};

HjelpetekstBase.defaultProps = {
    className: undefined
};

export default HjelpetekstBase;

export const HjelpetekstOver = (props) => <HjelpetekstBase type="over" {...props} />;
export const HjelpetekstUnder = (props) => <HjelpetekstBase type="under" {...props} />;
export const HjelpetekstHojre = (props) => <HjelpetekstBase type="hojre" {...props} />;
export const HjelpetekstVenstre = (props) => <HjelpetekstBase type="venstre" {...props} />;
