// nav-frontend-veileder/src/veileder-chat-bubble.tsx
import * as React from 'react';
import * as PT from 'prop-types';
import { StatelessComponent } from 'react';

export interface VeilederChatBubbleProps {
    says: React.ReactNode | React.ReactChild | React.ReactChildren;
    offset: number;
}

const VeilederChatBubble: StatelessComponent<VeilederChatBubbleProps> = (props) => {
    const { says, offset } = props;

    const scaleMultiplier = Math.max(0.5, offset / 150);

    const bubbleStyles = {
        bottom: offset + ((20 * scaleMultiplier) / 2),
        marginRight: offset / 2
    };

    const arrowStyles = {
        bottom: Math.ceil(-20 - (((20 * scaleMultiplier) / 2) - 10)),
        right: ((20 - (20 * scaleMultiplier)) * -1) + (20 * scaleMultiplier),
        transform: `scale(${scaleMultiplier})`
    };

    return (
        <div className="nav-frontend-veileder__chatBubble" style={bubbleStyles}>
            {says}
            <i className="nav-frontend-veileder__chatBubbleArrow" style={arrowStyles} />
        </div>
    );
};

(VeilederChatBubble as any).propTypes = {
    says: PT.node.isRequired,
    offset: PT.number.isRequired
};

export default VeilederChatBubble;
