// nav-frontend-veileder/src/veileder-snakkeboble.tsx
import * as React from 'react';
import * as PT from 'prop-types';
import { StatelessComponent } from 'react';

export interface NavVeilederSnakkebobleProps {
    sier: React.ReactNode | React.ReactChild | React.ReactChildren;
    offset: number;
};

const NavVeilederSnakkeboble: StatelessComponent<NavVeilederSnakkebobleProps> = props => {
    const { sier, offset } = props;

    const scaleMultiplier = Math.max(0.5, offset / 150);

    const bubbleStyles = {
        bottom: offset + ((20 * scaleMultiplier) / 2),
        marginRight: offset / 2
    };

    const arrowStyles = {
        bottom: Math.ceil(-20 - (((20 * scaleMultiplier) / 2) - 10)),
        right: ((20 - (20 * scaleMultiplier)) * -1) + (20 * scaleMultiplier),
        transform: `scale(${scaleMultiplier})`
    };

    return (
        <div className="nav-veileder__snakkeboble" style={bubbleStyles}>
            {sier}
            <i className="nav-veileder__snakkeboblePil" style={arrowStyles} />
        </div>
    );
};

(NavVeilederSnakkeboble as any).propTypes = {
    sier: PT.node.isRequired,
    offset: PT.number.isRequired
};

export default NavVeilederSnakkeboble;
