// nav-frontend-inngang/_inngang.sample.js
import {
    AppInngang,
    DobbelIngressInngang,
    EnkelIngressInngang,
    NormalVertikalInngang,
    StorVertikalInngang
} from './';

import {
    createSampleData,
    newType
} from '../../../guideline-app/app/utils/sampling/sampleDataHelper';

const href = { href: '/#/components/innganger' };

const types = [
    newType(AppInngang, 'Inngang 1', null, {
        tittel: 'Gå til en applikasjon',
        ...href
    }),
    newType(DobbelIngressInngang, 'Inngang 2', null, {
        tittel: 'Her er en tittel',
        ingress1: 'Her er ingress 1',
        ingress2: 'Her er ingress 2',
        ikon: 'nav-ansatt',
        ...href
    }),
    newType(EnkelIngressInngang, 'Inngang 3', null, {
        tittel: 'Her er en tittel',
        ingress: 'Her er en ingress',
        ikon: 'nav-ansatt',
        ...href
    }),
    newType(NormalVertikalInngang, 'Inngang 4', null, {
        tittel: 'Her er en tittel',
        tekst: 'Og her har du en tekst',
        lesMerHref: '/#/components/innganger/lenkeeksempel/',
        ...href
    }),
    newType(NormalVertikalInngang, 'Inngang 5', null, {
        tittel: 'Her er en tittel',
        tekst: 'Og her har du en tekst',
        ...href
    }),
    newType(NormalVertikalInngang, 'Inngang 6', null, {
        tittel: 'Her er en tittel',
        tekst: 'Og her har du en tekst',
        ikon: 'nav-ansatt',
        ...href
    }),
    newType(StorVertikalInngang, 'Inngang 7', null, {
        tittel: 'Her er en tittel',
        ingress: 'Og her har du en ingress',
        ikon: 'nav-ansatt',
        ...href
    })
];

export default createSampleData(types);
