// nav-frontend-skjema/sample/_bekreftcheckbokspanel.sample.no-transpilation.js
import React from 'react';
import { BekreftCheckboksPanel } from './../';
import { Checkbox } from "./..";

class BekreftCheckboksPanelExample extends React.Component {
    constructor(props) {
        super(props);
        this.state = { checked: false, feilmelding: false };
    }

    onChange(event) {
        // eslint-disable-next-line
        console.log('event', event);
        this.setState({ checked: !this.state.checked });
    }

	toggleFeil() {
		this.setState({ feilmelding: !this.state.feilmelding });
    }

    render() {
        return (
            <div>
                <BekreftCheckboksPanel
                    label="Jeg bekrefter at opplysningene jeg har gitt er riktige."
                    checked={this.state.checked}
                    onChange={(event) => this.onChange(event)}
                    feil={this.state.feilmelding ? {
                            feilmelding: "Du må bekrefte før søknaden kan sendes inn."
                        } : undefined}
                >
                    <span>
                        Jeg er kjent med at hvis opplysningene jeg har gitt ikke er riktige eller
                        fullstendige, så kan jeg miste retten til stønad.
                    </span>
                </BekreftCheckboksPanel>
                <br/>

                <Checkbox
                    label="Feilmelding"
                    onChange={() => this.toggleFeil()}
                />
            </div>
        );
    }
}

export default BekreftCheckboksPanelExample;
