// nav-frontend-lastestriper/src/lastestriper.tsx
import * as React from 'react';
import * as PT from 'prop-types';
import 'nav-frontend-lastestriper-style';

export interface NavFrontendLastestriperProps extends React.HTMLAttributes<HTMLDivElement> {
    /**
     * Antall lastestriper. Default er 3.
     */
    linjer: number;
}

class NavFrontendLastestriper extends React.Component<NavFrontendLastestriperProps, {}> {

    static defaultProps = {
        linjer: 3
    };

    lastestriper (linjer: number): React.ReactNode[] {
        return Array(linjer)
            .fill(0)
            .map((_, indeks) => {
                const idx = indeks % 3;
                if (idx === 0) {
                    return <div className="lastestripe" key={indeks} />;
                    // tslint:disable-next-line:no-else-after-return
                } else if (idx === 1) {
                    return <div className="lastestripe lastestripe__kort_forsinkelse" key={indeks} />;
                } else {
                    return <div className="lastestripe lastestripe__lang_forsinkelse lastestripe__smal" key={indeks} />;
                }
            });
    };

    render(): JSX.Element {
        const { linjer, style, className } = this.props;
        return (
            <div className={'nav-frontend-lastestriper' + (className ? ' ' + className : '')} style={style}>
                {this.lastestriper(linjer)}
            </div>
        );
    }
}

(NavFrontendLastestriper as React.ComponentClass).propTypes = {
    /**
     * Antall lastestriper. Default er 3.
     */
    linjer: PT.number,
    /**
     * Inline stiling
     */
    style: PT.object,
    /**
     * Egendefinert klassenavn.
     */
    className: PT.string
};

(NavFrontendLastestriper as React.ComponentClass).defaultProps = {
    linjer: 3
};

export default NavFrontendLastestriper;
