// nav-frontend-tekstomrade/src/tekstomrade.js
import PT from 'prop-types';
import React, { Component } from 'react';
import { Normaltekst } from 'nav-frontend-typografi';
import { SpanFragment, LinkFragment } from './fragments';

const uriRegex = /((?:[\w-]+:\/\/?|www(?:-\w+)?\.)[^\s()<>]+\w)/g;
const notNull = (element) => element !== null;

function tilJSXLenke(fragment, fragmentIndex) {
    const urimatch = uriRegex.exec(fragment);
    if (urimatch === null) {
        return <SpanFragment key={fragmentIndex}>{fragment}</SpanFragment>;
    }
    return <LinkFragment key={fragmentIndex} href={urimatch[0]} />;
}

function leggTilLenkerJSX(innhold) {
    const fragments = innhold.split(uriRegex);
    if (fragments.length === 1) {
        const text = fragments[0];
        if (!text || text.length === 0) {
            return null;
        }
        return (<SpanFragment>{text}</SpanFragment>);
    }

    return fragments
        .map(tilJSXLenke)
        .filter(notNull);
}
function tilAvsnitt(avsnitt, index, list) {
    return (
        <Normaltekst className={index < list.length - 1 ? 'blokk-xs' : ''} key={index}>{avsnitt}</Normaltekst>
    );
}

class Tekstomrade extends Component { // eslint-disable-line react/prefer-stateless-function
    render() {
        const { children, ingenFormattering, ...props } = this.props;

        const tekst = children;
        const avsnitt = ingenFormattering ? tekst : tekst.split(/[\r\n]+/)
            .map(leggTilLenkerJSX)
            .filter(notNull)
            .map(tilAvsnitt);

        return <div {...props}>{avsnitt}</div>;
    }
}

Tekstomrade.propTypes = {
    children: PT.string,
    ingenFormattering: PT.bool
};

Tekstomrade.defaultProps = {
    children: '',
    ingenFormattering: false
};

export default Tekstomrade;
