// nav-frontend-typografi/src/index.js
import React, { PropTypes as PT } from 'react';
import classNames from 'classnames';

export const typeConfigMap = {
    sidetittel: { defaultTag: 'h1', cls: (...args) => classNames('typo-sidetittel', ...args) },
    innholdstittel: { defaultTag: 'h2', cls: (...args) => classNames('typo-innholdstittel', ...args) },
    systemtittel: { defaultTag: 'h3', cls: (...args) => classNames('typo-systemtittel', ...args) },
    undertittel: { defaultTag: 'h3', cls: (...args) => classNames('typo-undertittel', ...args) },
    element: { defaultTag: 'p', cls: (...args) => classNames('typo-element', ...args) },
    avsnitt: { defaultTag: 'p', cls: (...args) => classNames('typo-avsnitt', ...args) },
    undertekst: { defaultTag: 'p', cls: (...args) => classNames('typo-undertekst', ...args) },
    infotekst: { defaultTag: 'p', cls: (...args) => classNames('typo-infotekst', ...args) },
    normal: { defaultTag: 'p', cls: (...args) => classNames('typo-normal', ...args) }
};

function getConfigForType(type) {
    const typeConfig = typeConfigMap[type];

    if (!typeConfig) {
        throw new Error(
            `Kunne ikke finne typeconfig for ${type}. Støttede typer er: ${JSON.stringify(Object.keys(typeConfigMap))}`
        );
    }

    return typeConfig;
}

function TypografiBase({ type, tag, className, children, ...props }) {
    const config = getConfigForType(type);
    const tagName = tag || config.defaultTag;

    const classnames = { className: config.cls(className) };

    return React.createElement(tagName, { ...props, ...classnames }, children);
}

TypografiBase.propTypes = {
    type: PT.string.isRequired,
    tag: PT.string, // eslint-disable-line react/require-default-props
    className: PT.string,
    children: PT.node.isRequired
};
TypografiBase.defaultProps = {
    className: undefined
};

export default TypografiBase;

export const Sidetittel = (props) => <TypografiBase type="sidetittel" {...props} />;
export const Innholdstittel = (props) => <TypografiBase type="innholdstittel" {...props} />;
export const Undertittel = (props) => <TypografiBase type="undertittel" {...props} />;
export const Systemtittel = (props) => <TypografiBase type="systemtittel" {...props} />;
export const Element = (props) => <TypografiBase type="element" {...props} />;
export const Avsnitt = (props) => <TypografiBase type="avsnitt" {...props} />;
export const Undertekst = (props) => <TypografiBase type="undertekst" {...props} />;
export const Infotekst = (props) => <TypografiBase type="infotekst" {...props} />;
export const Normaltekst = (props) => <TypografiBase type="normal" {...props} />;
