// nav-frontend-modal/_modal.example.js
import React, { Component } from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Knapp } from 'nav-frontend-knapper';
// eslint-disable-next-line import/no-extraneous-dependencies
import { Innholdstittel, Ingress, Normaltekst } from 'nav-frontend-typografi';
import Modal from './';

export default class ModalEksempel extends Component {
    componentWillMount() {
        this.state = {
            modalIsOpen: false
        };
    }

    openModal() {
        this.setState({ modalIsOpen: true });
    }

    closeModal() {
        this.setState({ modalIsOpen: false });
    }

    render() {
        return (
            <div>
                <Knapp onClick={() => this.openModal()}>Klikk for å åpne modal</Knapp>
                <Modal
                    isOpen={this.state.modalIsOpen}
                    onRequestClose={() => this.closeModal()}
                    closeButton={false}
                    contentLabel="Heisann."
                >
                    <Innholdstittel>Tittel i modalen</Innholdstittel>
                    <Ingress>Ingress i modalen</Ingress>
                    <Normaltekst>Normaltekst i modalen</Normaltekst>
                </Modal>
            </div>
        );
    }
}
