// nav-frontend-modal/modal-eksempel.js
import React, { Component } from 'react';
import { // eslint-disable-line import/no-extraneous-dependencies
    Innholdstittel,
    Normaltekst
} from 'nav-frontend-typografi';
import Modal from './src/index';

class ModalEksempel extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isOpen: false,
            closeButton: true,
            shouldCloseOnOverlayClick: true
        };

        this.handleOpen = this.handleOpen.bind(this);
        this.handleClose = this.handleClose.bind(this);
        this.handleOverlayCloseCheckbox = this.handleOverlayCloseCheckbox.bind(this);
        this.handleCloseButtonCheckbox = this.handleCloseButtonCheckbox.bind(this);
    }

    handleOverlayCloseCheckbox(e) {
        this.setState({ shouldCloseOnOverlayClick: e.target.checked });
    }

    handleCloseButtonCheckbox(e) {
        this.setState({ closeButton: e.target.checked });
    }

    handleOpen() {
        this.setState({ isOpen: true });
    }

    handleClose() {
        this.setState({ isOpen: false });
    }

    render() {
        const { isOpen, shouldCloseOnOverlayClick, closeButton } = this.state;
        return (
            <div>
                <Modal
                    className="begrenset-bredde"
                    contentLabel="contentlabel blir satt som aria-label"
                    isOpen={isOpen}
                    closeButton={closeButton}
                    onRequestClose={this.handleClose}
                    contentClass="contentClass"
                    shouldCloseOnOverlayClick={shouldCloseOnOverlayClick}
                >
                    <Innholdstittel tag="h1" style={{ paddingRight: '3rem' }}>Min modal</Innholdstittel>
                    <Normaltekst>
                        Lorem ipsum dolor sit amet, consectetur adipisicing elit.
                        Accusamus eum ex facilis inventore minus, molestiae mollitia reiciendis ut veniam voluptatem!
                        Earum, id, qui. Autem doloribus, hic natus odio quae tempore.
                    </Normaltekst>

                    <Normaltekst>
                        Lorem ipsum dolor sit amet, consectetur adipisicing elit.
                        Accusamus eum ex facilis inventore minus, molestiae mollitia reiciendis ut veniam voluptatem!
                        Earum, id, qui. Autem doloribus, hic natus odio quae tempore.
                    </Normaltekst>
                </Modal>
                <div>
                    <input
                        id="closeButtonCheckbox"
                        type="checkbox"
                        onChange={this.handleCloseButtonCheckbox}
                        checked={closeButton}
                    />
                    <label htmlFor="closeButtonCheckbox">closeButton</label>
                </div>
                <div>
                    <input
                        id="overlayCloseCheckbox"
                        type="checkbox"
                        onChange={this.handleOverlayCloseCheckbox}
                        checked={shouldCloseOnOverlayClick}
                    />
                    <label htmlFor="overlayCloseCheckbox">shouldCloseOnOverlayClick</label>
                </div>
                <button onClick={this.handleOpen}>Klikk for å åpne</button>
            </div>
        );
    }
}

export default ModalEksempel;
