// nav-frontend-alertstriper/src/index.js
import React, { PropTypes as PT } from 'react';
import classNames from 'classnames';
import { Infotekst } from 'nav-frontend-typografi';
import 'nav-frontend-alertstriper-style'; // eslint-disable-line import/extensions

const cls = (type, className) => classNames('alertstriper', className, {
    'alertstriper--advarsel--solid': type === 'advarsel-solid',
    'alertstriper--info--solid': type === 'info-solid',
    'alertstriper--info': type === 'info',
    'alertstriper--suksess': type === 'suksess',
    'alertstriper--suksess--solid': type === 'suksess-solid'
});

function Alertstriper({ type, className, children, ...props }) {
    return <div className={cls(type, className)} {...props}><Infotekst tag="span">{children}</Infotekst></div>;
}

Alertstriper.propTypes = {
    children: PT.node.isRequired,
    type: PT.oneOf(['advarsel-solid', 'suksess-solid', 'suksess', 'info', 'info-solid']).isRequired,
    className: PT.string
};

Alertstriper.defaultProps = {
    className: undefined
};

export default Alertstriper;

export const AlertStripeAdvarsel = (props) => <Alertstriper type="advarsel-solid" {...props} />;
export const AlertStripeSuksessSolid = (props) => <Alertstriper type="suksess-solid" {...props} />;
export const AlertStripeSuksess = (props) => <Alertstriper type="suksess" {...props} />;
export const AlertStripeInfoSolid = (props) => <Alertstriper type="info-solid" {...props} />;
export const AlertStripeInfo = (props) => <Alertstriper type="info" {...props} />;
