// nav-frontend-alertstriper/src/index.js
import React, { PropTypes as PT } from 'react';
import classNames from 'classnames';
import { Infotekst } from 'nav-frontend-typografi';
import Ikon from 'nav-frontend-ikoner-assets';
import 'nav-frontend-alertstriper-style';

const cls = (type, solid, className) => classNames('alertstripe', className, {
    'alertstripe--solid': solid,
    'alertstripe--advarsel': type === 'advarsel',
    'alertstripe--info': type === 'info',
    'alertstripe--suksess': type === 'suksess'
});
const ikonKind = (type, solid) => {
    const kindMap = {
        'ok-sirkel-fylt': type === 'suksess' && solid,
        'ok-sirkel': type === 'suksess' && !solid,
        'advarsel-trekant-fylt': type === 'advarsel' && solid,
        'info-sirkel-fylt': type === 'info' && solid,
        'info-sirkel-orange': type === 'info' && !solid
    };

    return Object.keys(kindMap)
        .map((key) => ({ key, predicate: kindMap[key] }))
        .find(({ key, predicate}) => predicate).key;
};


function Alertstriper({ type, solid, className, children, ...props }) {
    return (
        <div className={cls(type, solid, className)} {...props}>
            <span className="alertstripe__ikon">
                <Ikon kind={ikonKind(type, solid)} height="1.5rem" width="1.5rem" />
            </span>
            <Infotekst tag="span">{children}</Infotekst>
        </div>
    );
}

Alertstriper.propTypes = {
    children: PT.node.isRequired,
    type: PT.oneOf(['advarsel', 'suksess', 'info']).isRequired,
    solid: PT.bool,
    className: PT.string
};

Alertstriper.defaultProps = {
    className: undefined,
    solid: false
};

export default Alertstriper;

export const AlertStripeAdvarsel = (props) => <Alertstriper type="advarsel" solid {...props} />;
export const AlertStripeSuksessSolid = (props) => <Alertstriper type="suksess" solid {...props} />;
export const AlertStripeSuksess = (props) => <Alertstriper type="suksess" {...props} />;
export const AlertStripeInfoSolid = (props) => <Alertstriper type="info" solid {...props} />;
export const AlertStripeInfo = (props) => <Alertstriper type="info" {...props} />;
